{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Original kaggle notebook: https://www.kaggle.com/danielwolffram/cord-19-create-dataframe\n",
    "\n",
    "- 原本取的檔名: edited_outer_on_pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some exploration and preprocessing to create one dataframe and export it as .csv.\n",
    "\n",
    "Adjusted to data from 2020-05-01."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "_cell_guid": "79c7e3d0-c299-4dcb-8224-4455121ee9b0",
    "_kg_hide-input": true,
    "_uuid": "d629ff2d2480ee46fbb7e2d37f6b5fab8052498a"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/dist-packages/tqdm/std.py:666: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', 100)\n",
    "\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()\n",
    "\n",
    "import regex as re\n",
    "import glob\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and Prepare Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To read the JSON files we follow [COVID EDA: Initial Exploration Tool](https://www.kaggle.com/ivanegapratama/covid-eda-initial-exploration-tool)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "cord_uid                           0\n",
       "sha                            14124\n",
       "source                             0\n",
       "title                            163\n",
       "doi                             4086\n",
       "pmcid                          11748\n",
       "pubmed_id                      18559\n",
       "license                            0\n",
       "abstract                       11130\n",
       "publish_time                       8\n",
       "authors                         2591\n",
       "journal                         7059\n",
       "Microsoft Academic Paper ID    58923\n",
       "WHO #Covidence                 58119\n",
       "arxiv_id                       59211\n",
       "has_pdf_parse                      0\n",
       "has_pmc_xml_parse                  0\n",
       "full_text_file                  8461\n",
       "url                              440\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14124\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zjufx4fo</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>PMC125340</td>\n",
       "      <td>11742998</td>\n",
       "      <td>unk</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...</td>\n",
       "      <td>2001-12-17</td>\n",
       "      <td>Pasternak, Alexander O.; van den Born, Erwin; Spaan, Willy J.M.; Snijder, Eric J.</td>\n",
       "      <td>The EMBO Journal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125340?pdf=render</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ymceytj3</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>PMC125375</td>\n",
       "      <td>11980704</td>\n",
       "      <td>unk</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...</td>\n",
       "      <td>2002-05-01</td>\n",
       "      <td>Tan, Kemin; Zelus, Bruce D.; Meijers, Rob; Liu, Jin-huan; Bergelson, Jeffrey M.; Duke, Norma; Zh...</td>\n",
       "      <td>The EMBO Journal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125375?pdf=render</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha source  \\\n",
       "0  zjufx4fo  b2897e1277f56641193a6db73825f707eed3e4c9    PMC   \n",
       "1  ymceytj3  e3d0d482ebd9a8ba81c254cc433f314142e72174    PMC   \n",
       "\n",
       "                                                                                                 title  \\\n",
       "0  Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...   \n",
       "1                 Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family   \n",
       "\n",
       "                        doi      pmcid pubmed_id license  \\\n",
       "0  10.1093/emboj/20.24.7220  PMC125340  11742998     unk   \n",
       "1   10.1093/emboj/21.9.2076  PMC125375  11980704     unk   \n",
       "\n",
       "                                                                                              abstract  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...   \n",
       "\n",
       "  publish_time  \\\n",
       "0   2001-12-17   \n",
       "1   2002-05-01   \n",
       "\n",
       "                                                                                               authors  \\\n",
       "0                    Pasternak, Alexander O.; van den Born, Erwin; Spaan, Willy J.M.; Snijder, Eric J.   \n",
       "1  Tan, Kemin; Zelus, Bruce D.; Meijers, Rob; Liu, Jin-huan; Bergelson, Jeffrey M.; Duke, Norma; Zh...   \n",
       "\n",
       "            journal Microsoft Academic Paper ID WHO #Covidence arxiv_id  \\\n",
       "0  The EMBO Journal                         NaN            NaN      NaN   \n",
       "1  The EMBO Journal                         NaN            NaN      NaN   \n",
       "\n",
       "   has_pdf_parse  has_pmc_xml_parse  full_text_file  \\\n",
       "0           True               True  custom_license   \n",
       "1           True               True  custom_license   \n",
       "\n",
       "                                                  url  \n",
       "0  http://europepmc.org/articles/pmc125340?pdf=render  \n",
       "1  http://europepmc.org/articles/pmc125375?pdf=render  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_path = 'raw/'\n",
    "metadata_path = f'{root_path}metadata.csv'\n",
    "meta_df = pd.read_csv(metadata_path, dtype={\n",
    "        'pubmed_id': str,\n",
    "        'Microsoft Academic Paper ID': str, \n",
    "        'doi': str\n",
    "    }, \n",
    "    parse_dates=['publish_time'])\n",
    "meta_df = meta_df.rename(columns = {'source_x': 'source'})\n",
    "\n",
    "display(meta_df.isnull().sum())\n",
    "print(len(meta_df) - meta_df.has_pdf_parse.sum())\n",
    "meta_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # test\n",
    "# def highlight(col):\n",
    "#     '''\n",
    "#     highlight the maximum in a Series yellow.\n",
    "#     '''\n",
    "    \n",
    "#     indicator = color_indicator.loc[duplicated['cord_uid'], col.name]\n",
    "#     return ['background-color: yellow' if val else '' for val in indicator]\n",
    "\n",
    "# duplicated = meta_df[meta_df.duplicated('cord_uid', keep=False)]\n",
    "# color_indicator = duplicated.groupby('cord_uid').nunique(dropna=False) > 1\n",
    "\n",
    "# duplicated.style.apply(highlight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(59851, 45748, 58676)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df['cord_uid'].nunique(), \\\n",
    "meta_df['sha'].nunique(), \\\n",
    "meta_df['title'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(48410, 22851)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_json = glob.glob(f'{root_path}/**/pdf_json/*.json', recursive=True)\n",
    "all_json_pmc = glob.glob(f'{root_path}/**/pmc_json/*.json', recursive=True)\n",
    "\n",
    "len(all_json), len(all_json_pmc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pdf_json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C52\n",
      "c\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "all_possibilities = []\n",
    "for path in all_json[:100]:\n",
    "    with open(path) as f:\n",
    "        j = json.load(f)\n",
    "    for entry in j['body_text']:\n",
    "        all_possibilities.append(entry['section'])\n",
    "all_possibilities = list(set(all_possibilities))\n",
    "\n",
    "print(all_possibilities[29])\n",
    "print(re.sub('[^A-Za-z]', '', all_possibilities[29]).lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bf23011551b21af6538e43974e74e6078f538028: ... Human infection by avian influenza virus (AIV) subtypes H7 1,2 and H9N2 3 has been reported in various countries over the past few years. Recent research reveals that AIV subtype H7 shares some proper...\n"
     ]
    }
   ],
   "source": [
    "class FileReader:\n",
    "    def __init__(self, file_path):\n",
    "        with open(file_path) as file:\n",
    "            content = json.load(file)\n",
    "            self.paper_id = content['paper_id']\n",
    "            self.abstract = []\n",
    "            self.body_text = []\n",
    "            self.methods = []\n",
    "            self.results = []\n",
    "\n",
    "            # Abstract\n",
    "            for entry in content['abstract']:\n",
    "                self.abstract.append(entry['text'])\n",
    "                \n",
    "            # Body text\n",
    "            for entry in content['body_text']:\n",
    "                self.body_text.append(entry['text'])\n",
    "                \n",
    "            # methods\n",
    "# TODO: 這裡的對法怪怪的，可以再改過，現在就是先全部擠成一坨然後比對\n",
    "            methods = ['methods','method','statistical methods','materials','materials and methods',\n",
    "                'data collection','the study','study design','experimental design','objective',\n",
    "                'objectives','procedures','data collection and analysis', 'methodology',\n",
    "                'material and methods','the model','experimental procedures','main text']\n",
    "            methods = [re.sub('[^A-Za-z]', '', type_).lower() for type_ in methods]\n",
    "            \n",
    "            for entry in content['body_text']:\n",
    "                # remove numbers and spaces\n",
    "                # 'Fig. S12' -> 'figs'\n",
    "                section_title = re.sub('[^A-Za-z]', '', entry['section']).lower()\n",
    "                \n",
    "                if any(t in section_title for t in methods) : \n",
    "                    self.methods.append(entry['text'])\n",
    "                    \n",
    "            # Results\n",
    "            results_synonyms = ['result']\n",
    "            for entry in content['body_text']:\n",
    "                section_title = re.sub('[^A-Za-z]', '', entry['section']).lower()\n",
    "                if any(r in section_title for r in results_synonyms) :\n",
    "                    self.results.append(entry['text'])\n",
    "            \n",
    "            # seperate paragraphs\n",
    "            self.abstract = '\\n'.join(self.abstract)\n",
    "            self.body_text = '\\n'.join(self.body_text)\n",
    "            self.methods = '\\n'.join(self.methods)\n",
    "            self.results = '\\n'.join(self.results)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'{self.paper_id}: {self.abstract[:200]}... {self.body_text[:200]}...'\n",
    "\n",
    "# test\n",
    "first_row = FileReader(all_json[0])\n",
    "print(first_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing index: 0 of 48410\n",
      "Processing index: 4841 of 48410\n",
      "Processing index: 9682 of 48410\n",
      "Processing index: 14523 of 48410\n",
      "Processing index: 19364 of 48410\n",
      "Processing index: 24205 of 48410\n",
      "Processing index: 29046 of 48410\n",
      "Processing index: 33887 of 48410\n",
      "Processing index: 38728 of 48410\n",
      "Processing index: 43569 of 48410\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>abstract</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bf23011551b21af6538e43974e74e6078f538028</td>\n",
       "      <td></td>\n",
       "      <td>Human infection by avian influenza virus (AIV) subtypes H7 1,2 and H9N2 3 has been reported in v...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c07b523c9bf9af72f5b6ee8b586b13ef619bb3ca</td>\n",
       "      <td>Recent advances in the development and discovery of pharmacological interventions within the ubi...</td>\n",
       "      <td>The ubiquitin-proteasome system (UPS) controls the turnover and biological function of most prot...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>04908565dc76425cc6a3fabc1dff0bce3860e3c2</td>\n",
       "      <td>The history of agriculture includes many animal and plant disease events that have had major con...</td>\n",
       "      <td>An emerging risk to plant, animal and/or human health may be defined as a risk resulting from a ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c28ae565308bfabbd784a23fa065a59a49d8cfc3</td>\n",
       "      <td></td>\n",
       "      <td>Morbilliviruses belong to the order Mononegavirales, family Paramyxoviridae, and include measles...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>82e80640af0a98cc07c911c9476f693c5e9a62cf</td>\n",
       "      <td>Beijing has been one of the epicenters attacked most severely by the SARS-CoV (severe acute resp...</td>\n",
       "      <td>Accumulated number of probable cases and deaths of SARS in Beijing, China, and the world. Data s...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  bf23011551b21af6538e43974e74e6078f538028   \n",
       "1  c07b523c9bf9af72f5b6ee8b586b13ef619bb3ca   \n",
       "2  04908565dc76425cc6a3fabc1dff0bce3860e3c2   \n",
       "3  c28ae565308bfabbd784a23fa065a59a49d8cfc3   \n",
       "4  82e80640af0a98cc07c911c9476f693c5e9a62cf   \n",
       "\n",
       "                                                                                              abstract  \\\n",
       "0                                                                                                        \n",
       "1  Recent advances in the development and discovery of pharmacological interventions within the ubi...   \n",
       "2  The history of agriculture includes many animal and plant disease events that have had major con...   \n",
       "3                                                                                                        \n",
       "4  Beijing has been one of the epicenters attacked most severely by the SARS-CoV (severe acute resp...   \n",
       "\n",
       "                                                                                             body_text  \\\n",
       "0  Human infection by avian influenza virus (AIV) subtypes H7 1,2 and H9N2 3 has been reported in v...   \n",
       "1  The ubiquitin-proteasome system (UPS) controls the turnover and biological function of most prot...   \n",
       "2  An emerging risk to plant, animal and/or human health may be defined as a risk resulting from a ...   \n",
       "3  Morbilliviruses belong to the order Mononegavirales, family Paramyxoviridae, and include measles...   \n",
       "4  Accumulated number of probable cases and deaths of SARS in Beijing, China, and the world. Data s...   \n",
       "\n",
       "  methods results  \n",
       "0                  \n",
       "1                  \n",
       "2                  \n",
       "3                  \n",
       "4                  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_ = {'paper_id': [], 'abstract': [], 'body_text': [], 'methods': [], 'results': []}\n",
    "for idx, entry in enumerate(all_json):\n",
    "    if idx % (len(all_json) // 10) == 0:\n",
    "        print(f'Processing index: {idx} of {len(all_json)}')\n",
    "        \n",
    "    content = FileReader(entry)\n",
    "    dict_['paper_id'].append(content.paper_id)\n",
    "    dict_['abstract'].append(content.abstract)\n",
    "    dict_['body_text'].append(content.body_text)\n",
    "    dict_['methods'].append(content.methods)\n",
    "    dict_['results'].append(content.results)\n",
    "\n",
    "papers = pd.DataFrame(dict_, columns=['paper_id', 'abstract', 'body_text', 'methods', 'results'])\n",
    "papers.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34693, 37588)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test\n",
    "(papers.methods.str.len() == 0).sum(), (papers.results.str.len() == 0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18591, 5)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 找出 methods 或 results 的段落標題有符合的 paper\n",
    "indicator = (papers.results.str.len() != 0) | (papers.methods.str.len() != 0)\n",
    "papers[indicator].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.merge(meta_df, papers, left_on='sha', right_on='paper_id', how='right', suffixes=['', 'pdf']).drop('sha', axis=1)\n",
    "\n",
    "# print(df.columns)\n",
    "# print(df.shape)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cord_uid', 'source', 'title', 'doi', 'pmcid', 'pubmed_id', 'license',\n",
      "       'abstract', 'publish_time', 'authors', 'journal',\n",
      "       'Microsoft Academic Paper ID', 'WHO #Covidence', 'arxiv_id',\n",
      "       'has_pdf_parse', 'has_pmc_xml_parse', 'full_text_file', 'url',\n",
      "       'paper_id', 'abstract_pdf', 'body_text', 'methods', 'results'],\n",
      "      dtype='object')\n",
      "(64967, 23)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>...</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id</th>\n",
       "      <th>abstract_pdf</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zjufx4fo</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>PMC125340</td>\n",
       "      <td>11742998</td>\n",
       "      <td>unk</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...</td>\n",
       "      <td>2001-12-17</td>\n",
       "      <td>Pasternak, Alexander O.; van den Born, Erwin; Spaan, Willy J.M.; Snijder, Eric J.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125340?pdf=render</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5¢ end of the genome fused...</td>\n",
       "      <td>The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...</td>\n",
       "      <td>Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...</td>\n",
       "      <td>EAV genome replication is not signi®cantly affected by leader TRS and body TRS mutations To diss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ymceytj3</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>PMC125375</td>\n",
       "      <td>11980704</td>\n",
       "      <td>unk</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...</td>\n",
       "      <td>2002-05-01</td>\n",
       "      <td>Tan, Kemin; Zelus, Bruce D.; Meijers, Rob; Liu, Jin-huan; Bergelson, Jeffrey M.; Duke, Norma; Zh...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125375?pdf=render</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>contributed equally to this work CEACAM1 is a member of the carcinoembryonic antigen (CEA) famil...</td>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "      <td></td>\n",
       "      <td>Molecular structure of msCEACAM1a [1, 4] The msCEACAM1a[1,4] protein analyzed in this paper cont...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wzj2glte</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein by ribosomal frameshift</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "      <td>PMC125543</td>\n",
       "      <td>11447125</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...</td>\n",
       "      <td>2001-07-16</td>\n",
       "      <td>Xu, Zhenming; Choi, Jinah; Yen, T.S.Benedict; Lu, Wen; Strohecker, Anne; Govindarajan, Sugantha;...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/</td>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td></td>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...</td>\n",
       "      <td></td>\n",
       "      <td>Synthesis of a 17 kDa protein from a coding sequence that overlaps the core protein reading fram...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2sfqsfm1</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "      <td>PMC126080</td>\n",
       "      <td>12093723</td>\n",
       "      <td>unk</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...</td>\n",
       "      <td>2002-07-01</td>\n",
       "      <td>Anand, Kanchan; Palm, Gottfried J.; Mesters, Jeroen R.; Siddell, Stuart G.; Ziebuhr, John; Hilge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc126080?pdf=render</td>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M pro , a pro...</td>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "      <td>Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...</td>\n",
       "      <td>Structure determination by MAD phasing The presence of 10 methionine residues in the TGEV M pro ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i0zym7iq</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "      <td>PMC136939</td>\n",
       "      <td>12456663</td>\n",
       "      <td>unk</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...</td>\n",
       "      <td>2002-12-01</td>\n",
       "      <td>van Vliet, A.L.W.; Smits, S.L.; Rottier, P.J.M.; de Groot, R.J.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc136939?pdf=render</td>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>corona-, toro-and roniviruses are evolutionarily related positive-strand RNA viruses, united in ...</td>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "      <td>Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...</td>\n",
       "      <td>Toroviral mRNAs possess a 5 ¢ cap structure Although most RNA viruses provide their (+)strand tr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid source  \\\n",
       "0  zjufx4fo    PMC   \n",
       "1  ymceytj3    PMC   \n",
       "2  wzj2glte    PMC   \n",
       "3  2sfqsfm1    PMC   \n",
       "4  i0zym7iq    PMC   \n",
       "\n",
       "                                                                                                 title  \\\n",
       "0  Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...   \n",
       "1                 Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family   \n",
       "2                               Synthesis of a novel hepatitis C virus protein by ribosomal frameshift   \n",
       "3  Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...   \n",
       "4                      Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus   \n",
       "\n",
       "                        doi      pmcid pubmed_id license  \\\n",
       "0  10.1093/emboj/20.24.7220  PMC125340  11742998     unk   \n",
       "1   10.1093/emboj/21.9.2076  PMC125375  11980704     unk   \n",
       "2  10.1093/emboj/20.14.3840  PMC125543  11447125   no-cc   \n",
       "3      10.1093/emboj/cdf327  PMC126080  12093723     unk   \n",
       "4      10.1093/emboj/cdf635  PMC136939  12456663     unk   \n",
       "\n",
       "                                                                                              abstract  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...   \n",
       "2  Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...   \n",
       "3  The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...   \n",
       "4  Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...   \n",
       "\n",
       "  publish_time  \\\n",
       "0   2001-12-17   \n",
       "1   2002-05-01   \n",
       "2   2001-07-16   \n",
       "3   2002-07-01   \n",
       "4   2002-12-01   \n",
       "\n",
       "                                                                                               authors  \\\n",
       "0                    Pasternak, Alexander O.; van den Born, Erwin; Spaan, Willy J.M.; Snijder, Eric J.   \n",
       "1  Tan, Kemin; Zelus, Bruce D.; Meijers, Rob; Liu, Jin-huan; Bergelson, Jeffrey M.; Duke, Norma; Zh...   \n",
       "2  Xu, Zhenming; Choi, Jinah; Yen, T.S.Benedict; Lu, Wen; Strohecker, Anne; Govindarajan, Sugantha;...   \n",
       "3  Anand, Kanchan; Palm, Gottfried J.; Mesters, Jeroen R.; Siddell, Stuart G.; Ziebuhr, John; Hilge...   \n",
       "4                                      van Vliet, A.L.W.; Smits, S.L.; Rottier, P.J.M.; de Groot, R.J.   \n",
       "\n",
       "   ... arxiv_id has_pdf_parse has_pmc_xml_parse  full_text_file  \\\n",
       "0  ...      NaN          True              True  custom_license   \n",
       "1  ...      NaN          True              True  custom_license   \n",
       "2  ...      NaN          True              True  custom_license   \n",
       "3  ...      NaN          True              True  custom_license   \n",
       "4  ...      NaN          True              True  custom_license   \n",
       "\n",
       "                                                    url  \\\n",
       "0    http://europepmc.org/articles/pmc125340?pdf=render   \n",
       "1    http://europepmc.org/articles/pmc125375?pdf=render   \n",
       "2  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/   \n",
       "3    http://europepmc.org/articles/pmc126080?pdf=render   \n",
       "4    http://europepmc.org/articles/pmc136939?pdf=render   \n",
       "\n",
       "                                   paper_id  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                                                                          abstract_pdf  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5¢ end of the genome fused...   \n",
       "1  contributed equally to this work CEACAM1 is a member of the carcinoembryonic antigen (CEA) famil...   \n",
       "2                                                                                                        \n",
       "3  The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M pro , a pro...   \n",
       "4  corona-, toro-and roniviruses are evolutionarily related positive-strand RNA viruses, united in ...   \n",
       "\n",
       "                                                                                             body_text  \\\n",
       "0  The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...   \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...   \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...   \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...   \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...   \n",
       "\n",
       "                                                                                               methods  \\\n",
       "0  Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...   \n",
       "1                                                                                                        \n",
       "2                                                                                                        \n",
       "3  Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...   \n",
       "4  Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...   \n",
       "\n",
       "                                                                                               results  \n",
       "0  EAV genome replication is not signi®cantly affected by leader TRS and body TRS mutations To diss...  \n",
       "1  Molecular structure of msCEACAM1a [1, 4] The msCEACAM1a[1,4] protein analyzed in this paper cont...  \n",
       "2  Synthesis of a 17 kDa protein from a coding sequence that overlaps the core protein reading fram...  \n",
       "3  Structure determination by MAD phasing The presence of 10 methionine residues in the TGEV M pro ...  \n",
       "4  Toroviral mRNAs possess a 5 ¢ cap structure Although most RNA viruses provide their (+)strand tr...  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: 也許可以之後再做\n",
    "df_outer = pd.merge(meta_df, papers, left_on='sha', right_on='paper_id', how='outer', suffixes=['', '_pdf']).drop('sha', axis=1)\n",
    "\n",
    "print(df_outer.columns)\n",
    "print(df_outer.shape)\n",
    "df_outer.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pmc_json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This only contains the full text - no abstracts!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMC6810424: ...\n"
     ]
    }
   ],
   "source": [
    "# TODO: 把 pdf, pmc 的 FileReader 整合成一個 class\n",
    "class FileReader:\n",
    "    def __init__(self, file_path):\n",
    "        with open(file_path) as file:\n",
    "            content = json.load(file)\n",
    "            self.paper_id = content['paper_id']\n",
    "            self.body_text = []\n",
    "            self.methods = []\n",
    "            self.results = []\n",
    "\n",
    "            # Body text\n",
    "            for entry in content['body_text']:\n",
    "                self.body_text.append(entry['text'])\n",
    "            # Methods\n",
    "            methods = ['methods','method','statistical methods','materials','materials and methods',\n",
    "                'data collection','the study','study design','experimental design','objective',\n",
    "                'objectives','procedures','data collection and analysis', 'methodology',\n",
    "                'material and methods','the model','experimental procedures','main text']\n",
    "            for entry in content['body_text']:\n",
    "                section_title = ''.join(x.lower() for x in entry['section'] if x.isalpha()) #remove numbers and spaces\n",
    "                if any(m in section_title for m in [''.join(x.lower() for x in m if x.isalpha()) for m in methods]) : \n",
    "                    self.methods.append(entry['text'])\n",
    "            # Results\n",
    "            results_synonyms = ['result']\n",
    "            for entry in content['body_text']:\n",
    "                section_title = ''.join(x.lower() for x in entry['section'] if x.isalpha())\n",
    "                if any(r in section_title for r in results_synonyms) :\n",
    "                    self.results.append(entry['text'])\n",
    "                    \n",
    "            self.body_text = '\\n'.join(self.body_text)\n",
    "            self.methods = '\\n'.join(self.methods)\n",
    "            self.results = '\\n'.join(self.results)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'{self.paper_id}: {self.body_text[:200]}...'\n",
    "first_row = FileReader(all_json_pmc[0])\n",
    "print(first_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing index: 0 of 22851\n",
      "Processing index: 2285 of 22851\n",
      "Processing index: 4570 of 22851\n",
      "Processing index: 6855 of 22851\n",
      "Processing index: 9140 of 22851\n",
      "Processing index: 11425 of 22851\n",
      "Processing index: 13710 of 22851\n",
      "Processing index: 15995 of 22851\n",
      "Processing index: 18280 of 22851\n",
      "Processing index: 20565 of 22851\n",
      "Processing index: 22850 of 22851\n",
      "(22851, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC6810424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC2289574</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC4916766</td>\n",
       "      <td>Successful intracellular pathogens often use strategies to gain access to cellular compartments ...</td>\n",
       "      <td>For isolation of peripheral blood mononuclear cells (PBMCs), we obtained unidentified, discarded...</td>\n",
       "      <td>To determine whether MTb infection induces transcriptional activation of the IFITMs, we infected...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC5857709</td>\n",
       "      <td>The last decade has witnessed the emergence of several emerging infectious diseases, including t...</td>\n",
       "      <td>This survey was fielded in Singapore between November 2012 and February 2013 by a market researc...</td>\n",
       "      <td>\\nTable 3 presents demographic characteristics for the 500 respondents and for a comparable popu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC7031771</td>\n",
       "      <td>The on-going Coronavirus disease (COVID-19) outbreak in China has become the world's leading hea...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     paper_id  \\\n",
       "0  PMC6810424   \n",
       "1  PMC2289574   \n",
       "2  PMC4916766   \n",
       "3  PMC5857709   \n",
       "4  PMC7031771   \n",
       "\n",
       "                                                                                             body_text  \\\n",
       "0                                                                                                        \n",
       "1                                                                                                        \n",
       "2  Successful intracellular pathogens often use strategies to gain access to cellular compartments ...   \n",
       "3  The last decade has witnessed the emergence of several emerging infectious diseases, including t...   \n",
       "4  The on-going Coronavirus disease (COVID-19) outbreak in China has become the world's leading hea...   \n",
       "\n",
       "                                                                                               methods  \\\n",
       "0                                                                                                        \n",
       "1                                                                                                        \n",
       "2  For isolation of peripheral blood mononuclear cells (PBMCs), we obtained unidentified, discarded...   \n",
       "3  This survey was fielded in Singapore between November 2012 and February 2013 by a market researc...   \n",
       "4                                                                                                        \n",
       "\n",
       "                                                                                               results  \n",
       "0                                                                                                       \n",
       "1                                                                                                       \n",
       "2  To determine whether MTb infection induces transcriptional activation of the IFITMs, we infected...  \n",
       "3  \\nTable 3 presents demographic characteristics for the 500 respondents and for a comparable popu...  \n",
       "4                                                                                                       "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_ = {'paper_id': [], 'body_text': [], 'methods': [], 'results': []}\n",
    "for idx, entry in enumerate(all_json_pmc):\n",
    "    if idx % (len(all_json_pmc) // 10) == 0:\n",
    "        print(f'Processing index: {idx} of {len(all_json_pmc)}')\n",
    "    content = FileReader(entry)\n",
    "    dict_['paper_id'].append(content.paper_id)\n",
    "    dict_['body_text'].append(content.body_text)\n",
    "    dict_['methods'].append(content.methods)\n",
    "    dict_['results'].append(content.results)\n",
    "\n",
    "pmc_text = pd.DataFrame(dict_, columns=['paper_id', 'body_text', 'methods', 'results'])\n",
    "\n",
    "print(pmc_text.shape)\n",
    "pmc_text.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC6810424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC2289574</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PMC6252950</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PMC2119224</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PMC6735769</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      paper_id body_text methods results\n",
       "0   PMC6810424                          \n",
       "1   PMC2289574                          \n",
       "9   PMC6252950                          \n",
       "12  PMC2119224                          \n",
       "23  PMC6735769                          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22851, 4)\n",
      "after dropping empty body_texts...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [paper_id, body_text, methods, results]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20440, 4)\n"
     ]
    }
   ],
   "source": [
    "# Careful, some of the new texts are empty strings!\n",
    "display(pmc_text[pmc_text.body_text == ''].head())\n",
    "print(pmc_text.shape)\n",
    "\n",
    "print('after dropping empty body_texts...')\n",
    "\n",
    "pmc_text = pmc_text[pmc_text.body_text != '']\n",
    "display(pmc_text[pmc_text.body_text == ''].head())\n",
    "print(pmc_text.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48425, 28)\n",
      "Index(['cord_uid', 'sha', 'source', 'title', 'doi', 'pmcid', 'pubmed_id',\n",
      "       'license', 'abstract_meta', 'publish_time', 'authors', 'journal',\n",
      "       'Microsoft Academic Paper ID', 'WHO #Covidence', 'arxiv_id',\n",
      "       'has_pdf_parse', 'has_pmc_xml_parse', 'full_text_file', 'url',\n",
      "       'paper_id_pdf', 'abstract_pdf', 'body_text_pdf', 'methods_pdf',\n",
      "       'results_pdf', 'paper_id_pmc', 'body_text_pmc', 'methods_pmc',\n",
      "       'results_pmc'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract_meta</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>...</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id_pdf</th>\n",
       "      <th>abstract_pdf</th>\n",
       "      <th>body_text_pdf</th>\n",
       "      <th>methods_pdf</th>\n",
       "      <th>results_pdf</th>\n",
       "      <th>paper_id_pmc</th>\n",
       "      <th>body_text_pmc</th>\n",
       "      <th>methods_pmc</th>\n",
       "      <th>results_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zjufx4fo</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>PMC125340</td>\n",
       "      <td>11742998</td>\n",
       "      <td>unk</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...</td>\n",
       "      <td>2001-12-17</td>\n",
       "      <td>...</td>\n",
       "      <td>http://europepmc.org/articles/pmc125340?pdf=render</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5¢ end of the genome fused...</td>\n",
       "      <td>The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...</td>\n",
       "      <td>Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...</td>\n",
       "      <td>EAV genome replication is not signi®cantly affected by leader TRS and body TRS mutations To diss...</td>\n",
       "      <td>PMC125340</td>\n",
       "      <td>The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...</td>\n",
       "      <td>Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...</td>\n",
       "      <td>To dissect EAV RNA synthesis, we routinely use a full-length cDNA clone (van Dinten et al., 1997...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ymceytj3</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>PMC125375</td>\n",
       "      <td>11980704</td>\n",
       "      <td>unk</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...</td>\n",
       "      <td>2002-05-01</td>\n",
       "      <td>...</td>\n",
       "      <td>http://europepmc.org/articles/pmc125375?pdf=render</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>contributed equally to this work CEACAM1 is a member of the carcinoembryonic antigen (CEA) famil...</td>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "      <td></td>\n",
       "      <td>Molecular structure of msCEACAM1a [1, 4] The msCEACAM1a[1,4] protein analyzed in this paper cont...</td>\n",
       "      <td>PMC125375</td>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "      <td>Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...</td>\n",
       "      <td>The msCEACAM1a[1,4] protein analyzed in this paper contains the 202 extracellular amino acids of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wzj2glte</td>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein by ribosomal frameshift</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "      <td>PMC125543</td>\n",
       "      <td>11447125</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...</td>\n",
       "      <td>2001-07-16</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/</td>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td></td>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...</td>\n",
       "      <td></td>\n",
       "      <td>Synthesis of a 17 kDa protein from a coding sequence that overlaps the core protein reading fram...</td>\n",
       "      <td>PMC125543</td>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...</td>\n",
       "      <td>For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...</td>\n",
       "      <td>An inspection of the HCV-1 core protein-coding sequence revealed an overlapping coding sequence ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2sfqsfm1</td>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "      <td>PMC126080</td>\n",
       "      <td>12093723</td>\n",
       "      <td>unk</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...</td>\n",
       "      <td>2002-07-01</td>\n",
       "      <td>...</td>\n",
       "      <td>http://europepmc.org/articles/pmc126080?pdf=render</td>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M pro , a pro...</td>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "      <td>Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...</td>\n",
       "      <td>Structure determination by MAD phasing The presence of 10 methionine residues in the TGEV M pro ...</td>\n",
       "      <td>PMC126080</td>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "      <td>All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...</td>\n",
       "      <td>The presence of 10 methionine residues in the TGEV Mpro molecule suggested that selenomethionine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i0zym7iq</td>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "      <td>PMC136939</td>\n",
       "      <td>12456663</td>\n",
       "      <td>unk</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...</td>\n",
       "      <td>2002-12-01</td>\n",
       "      <td>...</td>\n",
       "      <td>http://europepmc.org/articles/pmc136939?pdf=render</td>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>corona-, toro-and roniviruses are evolutionarily related positive-strand RNA viruses, united in ...</td>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "      <td>Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...</td>\n",
       "      <td>Toroviral mRNAs possess a 5 ¢ cap structure Although most RNA viruses provide their (+)strand tr...</td>\n",
       "      <td>PMC136939</td>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "      <td>BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...</td>\n",
       "      <td>Although most RNA viruses provide their (+)strand transcripts with a 5′ cap (Furuichi and Shatki...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha source  \\\n",
       "0  zjufx4fo  b2897e1277f56641193a6db73825f707eed3e4c9    PMC   \n",
       "1  ymceytj3  e3d0d482ebd9a8ba81c254cc433f314142e72174    PMC   \n",
       "2  wzj2glte  00b1d99e70f779eb4ede50059db469c65e8c1469    PMC   \n",
       "3  2sfqsfm1  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0    PMC   \n",
       "4  i0zym7iq  dde02f11923815e6a16a31dd6298c46b109c5dfa    PMC   \n",
       "\n",
       "                                                                                                 title  \\\n",
       "0  Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...   \n",
       "1                 Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family   \n",
       "2                               Synthesis of a novel hepatitis C virus protein by ribosomal frameshift   \n",
       "3  Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...   \n",
       "4                      Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus   \n",
       "\n",
       "                        doi      pmcid pubmed_id license  \\\n",
       "0  10.1093/emboj/20.24.7220  PMC125340  11742998     unk   \n",
       "1   10.1093/emboj/21.9.2076  PMC125375  11980704     unk   \n",
       "2  10.1093/emboj/20.14.3840  PMC125543  11447125   no-cc   \n",
       "3      10.1093/emboj/cdf327  PMC126080  12093723     unk   \n",
       "4      10.1093/emboj/cdf635  PMC136939  12456663     unk   \n",
       "\n",
       "                                                                                         abstract_meta  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...   \n",
       "2  Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...   \n",
       "3  The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...   \n",
       "4  Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...   \n",
       "\n",
       "  publish_time  ...                                                   url  \\\n",
       "0   2001-12-17  ...    http://europepmc.org/articles/pmc125340?pdf=render   \n",
       "1   2002-05-01  ...    http://europepmc.org/articles/pmc125375?pdf=render   \n",
       "2   2001-07-16  ...  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/   \n",
       "3   2002-07-01  ...    http://europepmc.org/articles/pmc126080?pdf=render   \n",
       "4   2002-12-01  ...    http://europepmc.org/articles/pmc136939?pdf=render   \n",
       "\n",
       "                               paper_id_pdf  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                                                                          abstract_pdf  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5¢ end of the genome fused...   \n",
       "1  contributed equally to this work CEACAM1 is a member of the carcinoembryonic antigen (CEA) famil...   \n",
       "2                                                                                                        \n",
       "3  The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M pro , a pro...   \n",
       "4  corona-, toro-and roniviruses are evolutionarily related positive-strand RNA viruses, united in ...   \n",
       "\n",
       "                                                                                         body_text_pdf  \\\n",
       "0  The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...   \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...   \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...   \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...   \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...   \n",
       "\n",
       "                                                                                           methods_pdf  \\\n",
       "0  Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...   \n",
       "1                                                                                                        \n",
       "2                                                                                                        \n",
       "3  Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...   \n",
       "4  Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...   \n",
       "\n",
       "                                                                                           results_pdf  \\\n",
       "0  EAV genome replication is not signi®cantly affected by leader TRS and body TRS mutations To diss...   \n",
       "1  Molecular structure of msCEACAM1a [1, 4] The msCEACAM1a[1,4] protein analyzed in this paper cont...   \n",
       "2  Synthesis of a 17 kDa protein from a coding sequence that overlaps the core protein reading fram...   \n",
       "3  Structure determination by MAD phasing The presence of 10 methionine residues in the TGEV M pro ...   \n",
       "4  Toroviral mRNAs possess a 5 ¢ cap structure Although most RNA viruses provide their (+)strand tr...   \n",
       "\n",
       "  paper_id_pmc  \\\n",
       "0    PMC125340   \n",
       "1    PMC125375   \n",
       "2    PMC125543   \n",
       "3    PMC126080   \n",
       "4    PMC136939   \n",
       "\n",
       "                                                                                         body_text_pmc  \\\n",
       "0  The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...   \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...   \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...   \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...   \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...   \n",
       "\n",
       "                                                                                           methods_pmc  \\\n",
       "0  Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...   \n",
       "1  Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...   \n",
       "2  For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...   \n",
       "3  All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...   \n",
       "4  BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...   \n",
       "\n",
       "                                                                                           results_pmc  \n",
       "0  To dissect EAV RNA synthesis, we routinely use a full-length cDNA clone (van Dinten et al., 1997...  \n",
       "1  The msCEACAM1a[1,4] protein analyzed in this paper contains the 202 extracellular amino acids of...  \n",
       "2  An inspection of the HCV-1 core protein-coding sequence revealed an overlapping coding sequence ...  \n",
       "3  The presence of 10 methionine residues in the TGEV Mpro molecule suggested that selenomethionine...  \n",
       "4  Although most RNA viruses provide their (+)strand transcripts with a 5′ cap (Furuichi and Shatki...  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge meta_df, pdfs and pmcs\n",
    "\n",
    "# TODO: 原本是 outer 在 pdf, 然後 pmc，如果都 outer 在 meta 上，後面會有問題\n",
    "df = pd.merge(meta_df, papers, left_on='sha', right_on='paper_id', how='right', suffixes=['_meta', '_pdf'])\n",
    "df = pd.merge(df, pmc_text, left_on='pmcid', right_on='paper_id', how='left', suffixes=['_pdf', '_pmc'])\n",
    "\n",
    "# 丟掉被用來 merge 的 columns\n",
    "# df = df.drop(['pmcid', 'paper_id_pdf', 'paper_id_pmc'], axis=1)\n",
    "# 丟掉其他似乎沒用的 columns\n",
    "# df = df.drop(['has_pdf_parse', 'has_pmc_xml_parse', 'Microsoft Academic Paper ID', 'WHO #Covidence']\n",
    "\n",
    "print(df.shape)\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>abstract_pdf</th>\n",
       "      <th>body_text_pdf</th>\n",
       "      <th>methods_pdf</th>\n",
       "      <th>results_pdf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [has_pdf_parse, abstract_pdf, body_text_pdf, methods_pdf, results_pdf]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>body_text_pmc</th>\n",
       "      <th>methods_pmc</th>\n",
       "      <th>results_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    has_pmc_xml_parse body_text_pmc methods_pmc results_pmc\n",
       "91               True           NaN         NaN         NaN\n",
       "92               True           NaN         NaN         NaN\n",
       "270              True           NaN         NaN         NaN\n",
       "439              True           NaN         NaN         NaN\n",
       "444              True           NaN         NaN         NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2178\n"
     ]
    }
   ],
   "source": [
    "# has_pdf_parse but no corresponding jsons\n",
    "indicator = df['has_pdf_parse'] == True & df['body_text_pdf'].isna()\n",
    "cols = ['has_pdf_parse', 'abstract_pdf', 'body_text_pdf', 'methods_pdf', 'results_pdf']\n",
    "display(df[indicator][cols].head())\n",
    "print(indicator.sum())\n",
    "\n",
    "# has_pmc_parse but no corresponding jsons\n",
    "indicator = df['has_pmc_xml_parse'] == True & df['body_text_pmc'].isna()\n",
    "cols = ['has_pmc_xml_parse', 'body_text_pmc', 'methods_pmc', 'results_pmc']\n",
    "display(df[indicator][cols].head())\n",
    "print(indicator.sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploration/Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Different Abstract in Metadata and JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(44678, 28)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract_meta</th>\n",
       "      <th>abstract_pdf</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48415</th>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48416</th>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48417</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycyrrhetic acid polyglycosides were synthesized in one-pot via cationic ring-opening condensat...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48418</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A purified recombinant spike (S) protein was studied for its effect on stimulating human periphe...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48419</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Although susceptibility of neurons in the brain to microbial infection is a major determinant of...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48420</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A number of features were included in the final models, including article counts, indicator coun...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48421</th>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48422</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Background-Since the turn of the century seven new respiratory viruses have infected man and two...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48423</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Genetic factors are presumed to influence the incidence of feline infectious peritonitis (FIP), ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48424</th>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      abstract_meta  \\\n",
       "48415           NaN   \n",
       "48416           NaN   \n",
       "48417           NaN   \n",
       "48418           NaN   \n",
       "48419           NaN   \n",
       "48420           NaN   \n",
       "48421           NaN   \n",
       "48422           NaN   \n",
       "48423           NaN   \n",
       "48424           NaN   \n",
       "\n",
       "                                                                                              abstract_pdf  \\\n",
       "48415                                                                                                        \n",
       "48416                                                                                                        \n",
       "48417  Glycyrrhetic acid polyglycosides were synthesized in one-pot via cationic ring-opening condensat...   \n",
       "48418  A purified recombinant spike (S) protein was studied for its effect on stimulating human periphe...   \n",
       "48419  Although susceptibility of neurons in the brain to microbial infection is a major determinant of...   \n",
       "48420  A number of features were included in the final models, including article counts, indicator coun...   \n",
       "48421                                                                                                        \n",
       "48422  Background-Since the turn of the century seven new respiratory viruses have infected man and two...   \n",
       "48423  Genetic factors are presumed to influence the incidence of feline infectious peritonitis (FIP), ...   \n",
       "48424                                                                                                        \n",
       "\n",
       "       url  \n",
       "48415  NaN  \n",
       "48416  NaN  \n",
       "48417  NaN  \n",
       "48418  NaN  \n",
       "48419  NaN  \n",
       "48420  NaN  \n",
       "48421  NaN  \n",
       "48422  NaN  \n",
       "48423  NaN  \n",
       "48424  NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "different_abstracts = df[df.abstract_meta != df.abstract_pdf]\n",
    "\n",
    "print(different_abstracts.shape)\n",
    "with pd.option_context('display.max_colwidth', 100):\n",
    "    display(different_abstracts[['abstract_meta', 'abstract_pdf', 'url']].tail(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking some of the files online, it seems that where the abstract is missing in the metadata, the abstract in the JSON file is simply the beginning of the text.\n",
    "\n",
    "下列以一文章為例。\n",
    "\n",
    "## Changes of CT findings in a 2019 novel coronavirus (2019-nCoV) pneumonia patient\n",
    "於 all_json 的第 20023 篇，  \n",
    "url: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7107315/pdf/hcaa038.pdf  \n",
    "sha: 'b7555d21978c7989d2fe322f04f952f90b103d82'\n",
    "\n",
    "abstracts 為下列，可以發現標題不為 \"Abstracts\":\n",
    "### Learning points for clinicians\n",
    "The gold standard for diagnosis of 2019 novel coronavirus (2019-nCoV) infection is the new coronavirus\n",
    "nucleic acid in swabs, sputum, secretions from the lower\n",
    "respiratory tract or blood. However, due to the quite low\n",
    "sensitivity, short of supply and a relatively long period of\n",
    "new coronavirus nucleic acid kits, suspected cases with\n",
    "false-negative results remain a severe problem. Besides,\n",
    "a long incubation period (3–7 days, up to 14 days) and superior transmission capacity of new viruses contribute to\n",
    "the fast spread of 2019-nCoV. Under the circumstances,\n",
    "computed tomography imaging is not only useful for\n",
    "the detection, location of lesions but also helpful in\n",
    "the evaluation of the dynamic changes of patients with\n",
    "2019-nCoV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract_meta</th>\n",
       "      <th>abstract_pdf</th>\n",
       "      <th>url</th>\n",
       "      <th>body_text_pdf</th>\n",
       "      <th>body_text_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>58 ARDS = acute respiratory distress syndrome; HGMB1 = high mobility group B1; ICU = intensive c...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC420071/</td>\n",
       "      <td>John Marshall (University of Toronto) kicked off the meeting with an informative review of the e...</td>\n",
       "      <td>The eighth annual Toronto Critical Care Symposium was held from 30 October to 1 November 2003, i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>NaN</td>\n",
       "      <td>We agree with Doug Altman that correspondence is an essential part of medical publishing-acting ...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1140948/</td>\n",
       "      <td>Editors of medical journals accept that published research should be open to comment and correct...</td>\n",
       "      <td>In their attempt to defend heterosexual transmission as the driving force for HIV epidemics in s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Diversity' is now a magic word, showing up everywhere and often used to embrace many aspects of ...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1193560/</td>\n",
       "      <td>'Diversity' is now a magic word, showing up everywhere and often used to embrace many aspects of...</td>\n",
       "      <td>‘Diversity’ is now a magic word, showing up everywhere and often used to embrace many aspects of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>NaN</td>\n",
       "      <td>York, UK, was a good choice for the venue of this workshop because the history of nucleolar rese...</td>\n",
       "      <td>http://europepmc.org/articles/pmc1559673?pdf=render</td>\n",
       "      <td>We now have working catalogues of the components of human and plant (Arabidopsis) nucleoli. A fe...</td>\n",
       "      <td>We now have working catalogues of the components of human and plant (Arabidopsis) nucleoli. A fe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>NaN</td>\n",
       "      <td>for fi eld collections; J. Andre for research permits; T. Kreeger for access to his facility; R....</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2600301/</td>\n",
       "      <td>To the Editor: In the November 2007 issue of Emerging Infectious Diseases, Gayer et al. (1) desc...</td>\n",
       "      <td>In the November 2007 issue of Emerging Infectious Diseases, Gayer et al. (1) describe how confli...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    abstract_meta  \\\n",
       "23            NaN   \n",
       "79            NaN   \n",
       "103           NaN   \n",
       "209           NaN   \n",
       "607           NaN   \n",
       "\n",
       "                                                                                            abstract_pdf  \\\n",
       "23   58 ARDS = acute respiratory distress syndrome; HGMB1 = high mobility group B1; ICU = intensive c...   \n",
       "79   We agree with Doug Altman that correspondence is an essential part of medical publishing-acting ...   \n",
       "103  Diversity' is now a magic word, showing up everywhere and often used to embrace many aspects of ...   \n",
       "209  York, UK, was a good choice for the venue of this workshop because the history of nucleolar rese...   \n",
       "607  for fi eld collections; J. Andre for research permits; T. Kreeger for access to his facility; R....   \n",
       "\n",
       "                                                       url  \\\n",
       "23    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC420071/   \n",
       "79   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1140948/   \n",
       "103  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1193560/   \n",
       "209    http://europepmc.org/articles/pmc1559673?pdf=render   \n",
       "607  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2600301/   \n",
       "\n",
       "                                                                                           body_text_pdf  \\\n",
       "23   John Marshall (University of Toronto) kicked off the meeting with an informative review of the e...   \n",
       "79   Editors of medical journals accept that published research should be open to comment and correct...   \n",
       "103  'Diversity' is now a magic word, showing up everywhere and often used to embrace many aspects of...   \n",
       "209  We now have working catalogues of the components of human and plant (Arabidopsis) nucleoli. A fe...   \n",
       "607  To the Editor: In the November 2007 issue of Emerging Infectious Diseases, Gayer et al. (1) desc...   \n",
       "\n",
       "                                                                                           body_text_pmc  \n",
       "23   The eighth annual Toronto Critical Care Symposium was held from 30 October to 1 November 2003, i...  \n",
       "79   In their attempt to defend heterosexual transmission as the driving force for HIV epidemics in s...  \n",
       "103  ‘Diversity’ is now a magic word, showing up everywhere and often used to embrace many aspects of...  \n",
       "209  We now have working catalogues of the components of human and plant (Arabidopsis) nucleoli. A fe...  \n",
       "607  In the November 2007 issue of Emerging Infectious Diseases, Gayer et al. (1) describe how confli...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# json 有 abstract, 但 meta 沒有, 而且有網址的 data\n",
    "indicator = (df.abstract_meta.isnull()) & (df.abstract_pdf != '') & (~df.url.isnull())\n",
    "different_abstracts.loc[\n",
    "    indicator,\n",
    "    ['abstract_meta', 'abstract_pdf', 'url', 'body_text_pdf', 'body_text_pmc']\n",
    "].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     null  empty\n",
      "json 0 13737\n",
      "meta 10993 0\n"
     ]
    }
   ],
   "source": [
    "# TODO: 這邊數字怪怪的, 左上原本應為 0, 可能跟 dataframe merge 的方法有關\n",
    "print('     null  empty')\n",
    "\n",
    "# missing abstracts in json files\n",
    "print('json', df['abstract_pdf'].isnull().sum(), (df['abstract_pdf'] =='').sum())\n",
    "\n",
    "# missing abstracts in metadata\n",
    "print('meta', df['abstract_meta'].isnull().sum(), (df['abstract_meta']=='').sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the abstracts from the metadata seem more reliable we generally use these, but fill the missing values with the abstract from the extracted values from the JSON file.\n",
    "\n",
    "可能因為 json 中的 abstract，它的標題並不一定為 abstract 的原因。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6515"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indicator = (df.abstract_meta.isnull()) & (df.abstract_pdf != '')\n",
    "df.loc[indicator, 'abstract_meta'] = df.loc[indicator, 'abstract_pdf']\n",
    "\n",
    "df['abstract_meta'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the remaining missing values are also empty in the json files\n",
    "# TODO: 可能也因為 dataframe merge 的方法，這邊原本應該為 0\n",
    "\n",
    "(df.abstract_meta.isnull() & (df.abstract_pdf!='')).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cord_uid', 'sha', 'source', 'title', 'doi', 'pmcid', 'pubmed_id',\n",
      "       'license', 'abstract', 'publish_time', 'authors', 'journal',\n",
      "       'Microsoft Academic Paper ID', 'WHO #Covidence', 'arxiv_id',\n",
      "       'has_pdf_parse', 'has_pmc_xml_parse', 'full_text_file', 'url',\n",
      "       'paper_id_pdf', 'body_text_pdf', 'methods_pdf', 'results_pdf',\n",
      "       'paper_id_pmc', 'body_text_pmc', 'methods_pmc', 'results_pmc'],\n",
      "      dtype='object')\n",
      "(48425, 27)\n"
     ]
    }
   ],
   "source": [
    "df = df.rename(columns = {'abstract_meta': 'abstract'})\n",
    "df = df.drop('abstract_pdf', axis=1)\n",
    "\n",
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We still have to compare the text body from pdf and pmc files.\n",
    "\n",
    "# Quick comparison of both texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48425 (48425, 27)\n"
     ]
    }
   ],
   "source": [
    "print((df.body_text_pdf != df.body_text_pmc).sum(), df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>body_text_pdf</th>\n",
       "      <th>body_text_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...</td>\n",
       "      <td>The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...</td>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rhinoviruses have been associated with 40% to 65% of \"common colds\" studied by methods of virus ...</td>\n",
       "      <td>Rhinoviruses have been associated with 40% to 65% of \"common colds\" studied by methods of virus ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Charles Darwin recognized that the distribution and form of parasites was evolutionarily signifi...</td>\n",
       "      <td>Charles Darwin recognized that the distribution and form of parasites was evolutionarily signifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>The elongation phase of protein synthesis is a precise process and mechanisms exist to promote t...</td>\n",
       "      <td>The elongation phase of protein synthesis is a precise process and mechanisms exist to promote t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Salmonella enterica serovar Typhimurium (S.typhimurium) is a bacterial pathogen that invades and...</td>\n",
       "      <td>Salmonella enterica serovar Typhimurium (S.typhi murium) is a bacterial pathogen that invades an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>patient group, a further significant increase of the allele was seen in the \"Severe cases\" patie...</td>\n",
       "      <td>In late April 2003, Taiwan was stroked by an epidemic of severe acute respiratory syndrome (SARS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                         body_text_pdf  \\\n",
       "0  The genetic information of RNA viruses is organized very ef®ciently. Practically every nucleotid...   \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...   \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection...   \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...   \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...   \n",
       "5  Rhinoviruses have been associated with 40% to 65% of \"common colds\" studied by methods of virus ...   \n",
       "6  Charles Darwin recognized that the distribution and form of parasites was evolutionarily signifi...   \n",
       "7  The elongation phase of protein synthesis is a precise process and mechanisms exist to promote t...   \n",
       "8  Salmonella enterica serovar Typhimurium (S.typhimurium) is a bacterial pathogen that invades and...   \n",
       "9  patient group, a further significant increase of the allele was seen in the \"Severe cases\" patie...   \n",
       "\n",
       "                                                                                         body_text_pmc  \n",
       "0  The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...  \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...  \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...  \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...  \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...  \n",
       "5  Rhinoviruses have been associated with 40% to 65% of \"common colds\" studied by methods of virus ...  \n",
       "6  Charles Darwin recognized that the distribution and form of parasites was evolutionarily signifi...  \n",
       "7  The elongation phase of protein synthesis is a precise process and mechanisms exist to promote t...  \n",
       "8  Salmonella enterica serovar Typhimurium (S.typhi murium) is a bacterial pathogen that invades an...  \n",
       "9  In late April 2003, Taiwan was stroked by an epidemic of severe acute respiratory syndrome (SARS...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of~9.6 kb. Infection by this virus frequently leads to chronic infection, which in turn may lead to severe liver diseases including liver cirrhosis and hepatocellular carcinoma. The ®rst HCV genomic sequence, which was named the HCV-1 sequence, was isolated in 1989 (Choo et al., 1989) . Since then, many more HCV sequences have been reported. Based on their sequence similarities, these HCV sequences have been grouped into'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infection by this virus frequently leads to chronic infection, which in turn may lead to severe liver diseases including liver cirrhosis and hepatocellular carcinoma. The first HCV genomic sequence, which was named the HCV-1 sequence, was isolated in 1989 (Choo et al., 1989). Since then, many more HCV sequences have been reported. Based on their sequence similarities, these HCV sequences have been grouped int'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "indicator = (df.body_text_pdf != df.body_text_pmc) & df.body_text_pmc.notnull()\n",
    "display(df.loc[indicator, ['body_text_pdf', 'body_text_pmc']].head(10))\n",
    "\n",
    "display(df.loc[2, 'body_text_pdf'][:500])\n",
    "print()\n",
    "display(df.loc[2, 'body_text_pmc'][:500])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "可能擷取的段落不一樣，原因為 paper 格式不同，下面為另外一個例子，為原本 dataframe 的第 1242 筆 data:\n",
    "\n",
    "url:'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3100511/'\n",
    "\n",
    "### df.iloc[1242].body_text_pdf[:500]:  \n",
    "'The global prevalence of the dengue virus (DENV) has grown dramatically in recent decades, and it is now endemic in .100 countries, with some 2.5 billion people at risk of infection. Dengue is an arthropod-borne flavivirus that can be subdivided into the 4 major serotypes (DEN-1-DEN-4). Most dengue infections either are asymptomatic or lead to a self-limiting febrile illness, dengue fever. In some cases the illness is more severe, leading to dengue hemorrhagic fever (DHF) with severe plasma leak'\n",
    "    \n",
    "### df.iloc[1242].body_text_pmc[:500]:  \n",
    "'The dengue-2 strain, 16681, was grown in C6/36 cells, Vero cells, and monocyte-derived dendritic cells (DCs). Cell-free supernatants were used either neat or after concentration by ultracentrifugation at 45,000 rpm for 4 h at 4°C, and the virus pellet was resuspended in 1.5% fetal bovine serum (FBS)/Leibovitz L-15. To concentrate large volumes of low-titer DENV supernatant, DENV were precipitated with 10% Polyethylene glycol 8000 (Sigma) before ultracentrifugation. U937 were maintained in 10% FB'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Where available we use the text from the pmc file (body_text_pmc), trusting the statement that it is of higher quality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      pdf   pmc\n",
      "null  0 30109\n",
      "empty 0 0\n"
     ]
    }
   ],
   "source": [
    "# TODO: 跟 abstract 一樣怪怪的, 左上原本應為 0, 可能跟 dataframe merge 的方法有關\n",
    "print('      pdf   pmc')\n",
    "\n",
    "print('null ', df.body_text_pdf.isnull().sum(), df.body_text_pmc.isnull().sum())\n",
    "\n",
    "print('empty', (df.body_text_pdf == '').sum(), (df.body_text_pmc == '').sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "indicator = df.body_text_pmc.notnull()\n",
    "df.loc[indicator, 'body_text_pdf'] = df.loc[indicator, 'body_text_pmc']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: 可能也因為 dataframe merge 的方法，這邊原本應該為 0\n",
    "df['body_text_pdf'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cord_uid', 'sha', 'source', 'title', 'doi', 'pmcid', 'pubmed_id',\n",
      "       'license', 'abstract', 'publish_time', 'authors', 'journal',\n",
      "       'Microsoft Academic Paper ID', 'WHO #Covidence', 'arxiv_id',\n",
      "       'has_pdf_parse', 'has_pmc_xml_parse', 'full_text_file', 'url',\n",
      "       'paper_id_pdf', 'body_text', 'methods_pdf', 'results_pdf',\n",
      "       'paper_id_pmc', 'methods_pmc', 'results_pmc'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "df = df.rename(columns = {'body_text_pdf': 'body_text'})\n",
    "df = df.drop('body_text_pmc', axis=1)\n",
    "\n",
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>methods_pdf</th>\n",
       "      <th>methods_pmc</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...</td>\n",
       "      <td>Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...</td>\n",
       "      <td>http://europepmc.org/articles/pmc125340?pdf=render</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...</td>\n",
       "      <td>http://europepmc.org/articles/pmc125375?pdf=render</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...</td>\n",
       "      <td>All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...</td>\n",
       "      <td>http://europepmc.org/articles/pmc126080?pdf=render</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...</td>\n",
       "      <td>BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...</td>\n",
       "      <td>http://europepmc.org/articles/pmc136939?pdf=render</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           methods_pdf  \\\n",
       "0  Site-directed mutagenesis, RNA transfections and immuno¯uorescence analysis Site-directed mutage...   \n",
       "1                                                                                                        \n",
       "2                                                                                                        \n",
       "3  Protein puri®cation and crystallization Recombinant TGEV M pro was expressed and puri®ed as prev...   \n",
       "4  Cells and virus BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Colle...   \n",
       "\n",
       "                                                                                           methods_pmc  \\\n",
       "0  Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...   \n",
       "1  Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...   \n",
       "2  For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...   \n",
       "3  All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...   \n",
       "4  BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...   \n",
       "\n",
       "                                                    url  \n",
       "0    http://europepmc.org/articles/pmc125340?pdf=render  \n",
       "1    http://europepmc.org/articles/pmc125375?pdf=render  \n",
       "2  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/  \n",
       "3    http://europepmc.org/articles/pmc126080?pdf=render  \n",
       "4    http://europepmc.org/articles/pmc136939?pdf=render  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['methods_pmc'].notnull(), ['methods_pdf', 'methods_pmc', 'url']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    null  empty\n",
      "pdf 34708 0\n",
      "pmc 6622 30109\n"
     ]
    }
   ],
   "source": [
    "# TODO: 這邊數字怪怪的, 右上原本應為 0, 可能跟 dataframe merge 的方法有關\n",
    "print('    null  empty')\n",
    "print('pdf', (df['methods_pdf'] == '').sum(), df['methods_pdf'].isnull().sum())\n",
    "print('pmc', (df['methods_pmc'] == '').sum(), df['methods_pmc'].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    null  empty\n",
      "pdf 28791 0\n",
      "pmc 6622 30109\n",
      "Index(['cord_uid', 'sha', 'source', 'title', 'doi', 'pmcid', 'pubmed_id',\n",
      "       'license', 'abstract', 'publish_time', 'authors', 'journal',\n",
      "       'Microsoft Academic Paper ID', 'WHO #Covidence', 'arxiv_id',\n",
      "       'has_pdf_parse', 'has_pmc_xml_parse', 'full_text_file', 'url',\n",
      "       'paper_id_pdf', 'body_text', 'methods', 'results', 'paper_id_pmc'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>...</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id_pdf</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "      <th>paper_id_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zjufx4fo</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>PMC125340</td>\n",
       "      <td>11742998</td>\n",
       "      <td>unk</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...</td>\n",
       "      <td>2001-12-17</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125340?pdf=render</td>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...</td>\n",
       "      <td>Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...</td>\n",
       "      <td>To dissect EAV RNA synthesis, we routinely use a full-length cDNA clone (van Dinten et al., 1997...</td>\n",
       "      <td>PMC125340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ymceytj3</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>PMC125375</td>\n",
       "      <td>11980704</td>\n",
       "      <td>unk</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...</td>\n",
       "      <td>2002-05-01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc125375?pdf=render</td>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...</td>\n",
       "      <td>Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...</td>\n",
       "      <td>The msCEACAM1a[1,4] protein analyzed in this paper contains the 202 extracellular amino acids of...</td>\n",
       "      <td>PMC125375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wzj2glte</td>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein by ribosomal frameshift</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "      <td>PMC125543</td>\n",
       "      <td>11447125</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...</td>\n",
       "      <td>2001-07-16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/</td>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...</td>\n",
       "      <td>For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...</td>\n",
       "      <td>An inspection of the HCV-1 core protein-coding sequence revealed an overlapping coding sequence ...</td>\n",
       "      <td>PMC125543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2sfqsfm1</td>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "      <td>PMC126080</td>\n",
       "      <td>12093723</td>\n",
       "      <td>unk</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...</td>\n",
       "      <td>2002-07-01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc126080?pdf=render</td>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...</td>\n",
       "      <td>All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...</td>\n",
       "      <td>The presence of 10 methionine residues in the TGEV Mpro molecule suggested that selenomethionine...</td>\n",
       "      <td>PMC126080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i0zym7iq</td>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "      <td>PMC136939</td>\n",
       "      <td>12456663</td>\n",
       "      <td>unk</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...</td>\n",
       "      <td>2002-12-01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>http://europepmc.org/articles/pmc136939?pdf=render</td>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...</td>\n",
       "      <td>BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...</td>\n",
       "      <td>Although most RNA viruses provide their (+)strand transcripts with a 5′ cap (Furuichi and Shatki...</td>\n",
       "      <td>PMC136939</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha source  \\\n",
       "0  zjufx4fo  b2897e1277f56641193a6db73825f707eed3e4c9    PMC   \n",
       "1  ymceytj3  e3d0d482ebd9a8ba81c254cc433f314142e72174    PMC   \n",
       "2  wzj2glte  00b1d99e70f779eb4ede50059db469c65e8c1469    PMC   \n",
       "3  2sfqsfm1  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0    PMC   \n",
       "4  i0zym7iq  dde02f11923815e6a16a31dd6298c46b109c5dfa    PMC   \n",
       "\n",
       "                                                                                                 title  \\\n",
       "0  Sequence requirements for RNA strand transfer during nidovirus discontinuous subgenomic RNA synt...   \n",
       "1                 Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family   \n",
       "2                               Synthesis of a novel hepatitis C virus protein by ribosomal frameshift   \n",
       "3  Structure of coronavirus main proteinase reveals combination of a chymotrypsin fold with an extr...   \n",
       "4                      Discontinuous and non-discontinuous subgenomic RNA transcription in a nidovirus   \n",
       "\n",
       "                        doi      pmcid pubmed_id license  \\\n",
       "0  10.1093/emboj/20.24.7220  PMC125340  11742998     unk   \n",
       "1   10.1093/emboj/21.9.2076  PMC125375  11980704     unk   \n",
       "2  10.1093/emboj/20.14.3840  PMC125543  11447125   no-cc   \n",
       "3      10.1093/emboj/cdf327  PMC126080  12093723     unk   \n",
       "4      10.1093/emboj/cdf635  PMC136939  12456663     unk   \n",
       "\n",
       "                                                                                              abstract  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic antigen (CEA) family. Isoforms of murine CEACAM1 ser...   \n",
       "2  Hepatitis C virus (HCV) is an important human pathogen that affects ∼100 million people worldwid...   \n",
       "3  The key enzyme in coronavirus polyprotein processing is the viral main proteinase, M(pro), a pro...   \n",
       "4  Arteri-, corona-, toro- and roniviruses are evolutionarily related positive-strand RNA viruses, ...   \n",
       "\n",
       "  publish_time  ... arxiv_id has_pdf_parse has_pmc_xml_parse  full_text_file  \\\n",
       "0   2001-12-17  ...      NaN          True              True  custom_license   \n",
       "1   2002-05-01  ...      NaN          True              True  custom_license   \n",
       "2   2001-07-16  ...      NaN          True              True  custom_license   \n",
       "3   2002-07-01  ...      NaN          True              True  custom_license   \n",
       "4   2002-12-01  ...      NaN          True              True  custom_license   \n",
       "\n",
       "                                                    url  \\\n",
       "0    http://europepmc.org/articles/pmc125340?pdf=render   \n",
       "1    http://europepmc.org/articles/pmc125375?pdf=render   \n",
       "2  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC125543/   \n",
       "3    http://europepmc.org/articles/pmc126080?pdf=render   \n",
       "4    http://europepmc.org/articles/pmc136939?pdf=render   \n",
       "\n",
       "                               paper_id_pdf  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                                                                             body_text  \\\n",
       "0  The genetic information of RNA viruses is organized very efficiently. Practically every nucleoti...   \n",
       "1  Carcinoembryonic antigen (CEA; CD66e) was initially discovered as a tumor antigen (Gold and Free...   \n",
       "2  Hepatitis C virus (HCV) is a positive-stranded RNA virus with a genome size of ∼9.6 kb. Infectio...   \n",
       "3  Transmissible gastroenteritis virus (TGEV) belongs to the Coronaviridae, a family of positive-st...   \n",
       "4  Positive (+)-strand RNA viruses have developed a wide range of strategies to express their genes...   \n",
       "\n",
       "                                                                                               methods  \\\n",
       "0  Site-directed mutagenesis of EAV leader and body TRSs was carried out as described by van Marle ...   \n",
       "1  Nucleotide sequences encoding the first 236 amino acids of murine CEACAM1a[1,4], including the n...   \n",
       "2  For the construction of pCMV-CC, the HCV-1 core protein-coding sequence was isolated by PCR usin...   \n",
       "3  All six copies (designated A–F) of the TGEV Mpro in the asymmetric unit of the crystal could be ...   \n",
       "4  BEV was grown in equine dermis (Ederm) cells (American Type Tissue Culture Collection). A BEV st...   \n",
       "\n",
       "                                                                                               results  \\\n",
       "0  To dissect EAV RNA synthesis, we routinely use a full-length cDNA clone (van Dinten et al., 1997...   \n",
       "1  The msCEACAM1a[1,4] protein analyzed in this paper contains the 202 extracellular amino acids of...   \n",
       "2  An inspection of the HCV-1 core protein-coding sequence revealed an overlapping coding sequence ...   \n",
       "3  The presence of 10 methionine residues in the TGEV Mpro molecule suggested that selenomethionine...   \n",
       "4  Although most RNA viruses provide their (+)strand transcripts with a 5′ cap (Furuichi and Shatki...   \n",
       "\n",
       "  paper_id_pmc  \n",
       "0    PMC125340  \n",
       "1    PMC125375  \n",
       "2    PMC125543  \n",
       "3    PMC126080  \n",
       "4    PMC136939  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use methods_pmc when it's available\n",
    "indicator = (df.methods_pmc.notnull()) & (df.methods_pmc != '')\n",
    "df.loc[indicator, 'methods_pdf'] = df.loc[indicator, 'methods_pmc']\n",
    "\n",
    "# same for results\n",
    "indicator = (df.results_pmc.notnull()) & (df.results_pmc != '')\n",
    "df.loc[indicator, 'results_pdf'] = df.loc[indicator, 'results_pmc']\n",
    "\n",
    "# TODO: 這邊數字怪怪的, 右上原本應為 0, 可能跟 dataframe merge 的方法有關\n",
    "print('    null  empty')\n",
    "print('pdf', (df['methods_pdf'] == '').sum(), df['methods_pdf'].isnull().sum())\n",
    "print('pmc', (df['methods_pmc'] == '').sum(), df['methods_pmc'].isnull().sum())\n",
    "\n",
    "df.rename(columns = {'methods_pdf': 'methods', 'results_pdf': 'results'}, inplace=True)\n",
    "df.drop(columns=['methods_pmc', 'results_pmc'], inplace=True)\n",
    "\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43330, (48425, 24))"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Some paper ids are duplicated\n",
    "df['sha'].nunique(), df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sha</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31867</th>\n",
       "      <td>0ed3c6a5559cd73307184f51fc53ccc76da559bc</td>\n",
       "      <td>In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31866</th>\n",
       "      <td>0ed3c6a5559cd73307184f51fc53ccc76da559bc</td>\n",
       "      <td>In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31865</th>\n",
       "      <td>0ed3c6a5559cd73307184f51fc53ccc76da559bc</td>\n",
       "      <td>In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22633</th>\n",
       "      <td>45e40b2d7d973ed5c9798da613fb3cfa4427e2e2</td>\n",
       "      <td>A digestible vaccine sandwich VACCINE BIOTECHNOLOGY edited by James L. Bittle and Frederick L. M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22634</th>\n",
       "      <td>45e40b2d7d973ed5c9798da613fb3cfa4427e2e2</td>\n",
       "      <td>A digestible vaccine sandwich VACCINE BIOTECHNOLOGY edited by James L. Bittle and Frederick L. M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sha  \\\n",
       "31867  0ed3c6a5559cd73307184f51fc53ccc76da559bc   \n",
       "31866  0ed3c6a5559cd73307184f51fc53ccc76da559bc   \n",
       "31865  0ed3c6a5559cd73307184f51fc53ccc76da559bc   \n",
       "22633  45e40b2d7d973ed5c9798da613fb3cfa4427e2e2   \n",
       "22634  45e40b2d7d973ed5c9798da613fb3cfa4427e2e2   \n",
       "\n",
       "                                                                                                 body_text  \n",
       "31867  In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...  \n",
       "31866  In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...  \n",
       "31865  In this Journal, Zhu et al. recently reported the results of their genomic analysis of multidrug...  \n",
       "22633  A digestible vaccine sandwich VACCINE BIOTECHNOLOGY edited by James L. Bittle and Frederick L. M...  \n",
       "22634  A digestible vaccine sandwich VACCINE BIOTECHNOLOGY edited by James L. Bittle and Frederick L. M...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: sha 裡面會有 Nan\n",
    "duplicated = df.duplicated('sha', keep=False) & df['sha'].notna()\n",
    "print(duplicated.sum())\n",
    "df.loc[duplicated, ['sha', 'body_text']].sort_values('sha').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But luckily they also have the same text body. So we will just keep one article per paper_id.\n",
    "Check for example [https://www.sciencedirect.com/science/article/pii/S1386653209701295?via%3Dihub](https://www.sciencedirect.com/science/article/pii/S1386653209701295?via%3Dihub) and [https://www.sciencedirect.com/science/article/pii/S1386653209701325?via%3Dihub](https://www.sciencedirect.com/science/article/pii/S1386653209701325?via%3Dihub) - they have the same content."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(47884, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>...</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id_pdf</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "      <th>paper_id_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14210</th>\n",
       "      <td>4fbr8fx8</td>\n",
       "      <td>bd3945ec343cc4535650c9425aa143f71716de7c</td>\n",
       "      <td>PMC</td>\n",
       "      <td>The clinical importance of the nomenclature, evolution and taxonomy of human papillomaviruses</td>\n",
       "      <td>10.1016/j.jcv.2004.10.021</td>\n",
       "      <td>PMC7108213</td>\n",
       "      <td>15571999</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Human papillomaviruses (HPVs) are formally described by isolation of their circular double-stran...</td>\n",
       "      <td>2005-02-08</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7108213/</td>\n",
       "      <td>bd3945ec343cc4535650c9425aa143f71716de7c</td>\n",
       "      <td>Until the late 1970s, papillomaviruses attracted little interest as they were only known as the ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PMC7108213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14211</th>\n",
       "      <td>4fbr8fx8</td>\n",
       "      <td>bd3945ec343cc4535650c9425aa143f71716de7c</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>The clinical importance of the nomenclature, evolution and taxonomy of human papillomaviruses</td>\n",
       "      <td>10.1016/j.jcv.2004.10.021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15753006</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Human papillomaviruses (HPVs) are formally described by isolation of their circular dou...</td>\n",
       "      <td>2005-03-31</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.jcv.2004.10.021</td>\n",
       "      <td>bd3945ec343cc4535650c9425aa143f71716de7c</td>\n",
       "      <td>Until the late 1970s, papillomaviruses attracted little interest as they were only known as the ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid                                       sha    source  \\\n",
       "14210  4fbr8fx8  bd3945ec343cc4535650c9425aa143f71716de7c       PMC   \n",
       "14211  4fbr8fx8  bd3945ec343cc4535650c9425aa143f71716de7c  Elsevier   \n",
       "\n",
       "                                                                                               title  \\\n",
       "14210  The clinical importance of the nomenclature, evolution and taxonomy of human papillomaviruses   \n",
       "14211  The clinical importance of the nomenclature, evolution and taxonomy of human papillomaviruses   \n",
       "\n",
       "                             doi       pmcid pubmed_id    license  \\\n",
       "14210  10.1016/j.jcv.2004.10.021  PMC7108213  15571999      no-cc   \n",
       "14211  10.1016/j.jcv.2004.10.021         NaN  15753006  els-covid   \n",
       "\n",
       "                                                                                                  abstract  \\\n",
       "14210  Human papillomaviruses (HPVs) are formally described by isolation of their circular double-stran...   \n",
       "14211  Abstract Human papillomaviruses (HPVs) are formally described by isolation of their circular dou...   \n",
       "\n",
       "      publish_time  ... arxiv_id has_pdf_parse has_pmc_xml_parse  \\\n",
       "14210   2005-02-08  ...      NaN          True              True   \n",
       "14211   2005-03-31  ...      NaN          True             False   \n",
       "\n",
       "       full_text_file                                                    url  \\\n",
       "14210  custom_license  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7108213/   \n",
       "14211  custom_license              https://doi.org/10.1016/j.jcv.2004.10.021   \n",
       "\n",
       "                                   paper_id_pdf  \\\n",
       "14210  bd3945ec343cc4535650c9425aa143f71716de7c   \n",
       "14211  bd3945ec343cc4535650c9425aa143f71716de7c   \n",
       "\n",
       "                                                                                                 body_text  \\\n",
       "14210  Until the late 1970s, papillomaviruses attracted little interest as they were only known as the ...   \n",
       "14211  Until the late 1970s, papillomaviruses attracted little interest as they were only known as the ...   \n",
       "\n",
       "      methods results paper_id_pmc  \n",
       "14210                   PMC7108213  \n",
       "14211                          NaN  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.drop_duplicates(['sha', 'body_text'])    # sha, body_text 都一樣才會 drop\n",
    "print(df.shape)\n",
    "df[df.duplicated(['sha'], keep=False)].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>...</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id_pdf</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "      <th>paper_id_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22558</th>\n",
       "      <td>fjmchbew</td>\n",
       "      <td>58be092086c74c58e9067121a6ba4836468e7ec3</td>\n",
       "      <td>CZI</td>\n",
       "      <td>The Author's Response: Case of the Index Patient Who Caused Tertiary Transmission of Coronavirus...</td>\n",
       "      <td>10.3346/jkms.2020.35.e89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32080993</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>noncomm_use_subset</td>\n",
       "      <td>https://doi.org/10.3346/jkms.2020.35.e89</td>\n",
       "      <td>58be092086c74c58e9067121a6ba4836468e7ec3</td>\n",
       "      <td>administration reduces viral load. Also, the authors explain that LPV/r administration has also ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22559</th>\n",
       "      <td>u4taazr3</td>\n",
       "      <td>58be092086c74c58e9067121a6ba4836468e7ec3</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Letter to the Editor: Case of the Index Patient Who Caused Tertiary Transmission of Coronavirus ...</td>\n",
       "      <td>10.3346/jkms.2020.35.e88</td>\n",
       "      <td>PMC7036343</td>\n",
       "      <td>32080992</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-02-20</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>noncomm_use_subset</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7036343/</td>\n",
       "      <td>58be092086c74c58e9067121a6ba4836468e7ec3</td>\n",
       "      <td>I am grateful that Dr. Lim and his colleagues reported a case of COVID-19 that caused tertiary t...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PMC7036343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid                                       sha source  \\\n",
       "22558  fjmchbew  58be092086c74c58e9067121a6ba4836468e7ec3    CZI   \n",
       "22559  u4taazr3  58be092086c74c58e9067121a6ba4836468e7ec3    PMC   \n",
       "\n",
       "                                                                                                     title  \\\n",
       "22558  The Author's Response: Case of the Index Patient Who Caused Tertiary Transmission of Coronavirus...   \n",
       "22559  Letter to the Editor: Case of the Index Patient Who Caused Tertiary Transmission of Coronavirus ...   \n",
       "\n",
       "                            doi       pmcid pubmed_id   license abstract  \\\n",
       "22558  10.3346/jkms.2020.35.e89         NaN  32080993  cc-by-nc      NaN   \n",
       "22559  10.3346/jkms.2020.35.e88  PMC7036343  32080992  cc-by-nc      NaN   \n",
       "\n",
       "      publish_time  ... arxiv_id has_pdf_parse has_pmc_xml_parse  \\\n",
       "22558   2020-01-01  ...      NaN          True             False   \n",
       "22559   2020-02-20  ...      NaN          True              True   \n",
       "\n",
       "           full_text_file  \\\n",
       "22558  noncomm_use_subset   \n",
       "22559  noncomm_use_subset   \n",
       "\n",
       "                                                         url  \\\n",
       "22558               https://doi.org/10.3346/jkms.2020.35.e89   \n",
       "22559  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7036343/   \n",
       "\n",
       "                                   paper_id_pdf  \\\n",
       "22558  58be092086c74c58e9067121a6ba4836468e7ec3   \n",
       "22559  58be092086c74c58e9067121a6ba4836468e7ec3   \n",
       "\n",
       "                                                                                                 body_text  \\\n",
       "22558  administration reduces viral load. Also, the authors explain that LPV/r administration has also ...   \n",
       "22559  I am grateful that Dr. Lim and his colleagues reported a case of COVID-19 that caused tertiary t...   \n",
       "\n",
       "      methods results paper_id_pmc  \n",
       "22558                          NaN  \n",
       "22559                   PMC7036343  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: sha 裡面會有 Nan\n",
    "# TODO: 這裡並沒有特別處理 sha 一樣，但 body_text 不一樣的 data，直接 drop\n",
    "duplicated = df.duplicated('sha', keep=False) & df['sha'].notna()\n",
    "df[duplicated].sort_values('sha').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43330, (43331, 24))"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.drop_duplicates(['sha'])\n",
    "df['sha'].nunique(), df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>...</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "      <th>paper_id_pdf</th>\n",
       "      <th>body_text</th>\n",
       "      <th>methods</th>\n",
       "      <th>results</th>\n",
       "      <th>paper_id_pmc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21429</th>\n",
       "      <td>zp9k1k3z</td>\n",
       "      <td>aecbc613ebdab36753235197ffb4f35734b5ca63</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Clinical and immunologic studies in identical twins discordant for systemic lupus erythematosus</td>\n",
       "      <td>10.1016/0002-9343(73)90176-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4579077</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Middle-aged female identical twins, one of whom had systemic lupus erythematosus (SLE),...</td>\n",
       "      <td>1973-08-31</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(73)90176-9</td>\n",
       "      <td>aecbc613ebdab36753235197ffb4f35734b5ca63</td>\n",
       "      <td>The patient (Fo, ) was a 58 year old mentally retarded white woman, born in a rural area of sout...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21430</th>\n",
       "      <td>b10e023s</td>\n",
       "      <td>55b8337a50d270af02549897d75e22bea89b2083</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Epidemic viral gastroenteritis</td>\n",
       "      <td>10.1016/0002-9343(79)90457-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>222141</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Epidemic viral gastroenteritis is a significant world wide problem. In developed countr...</td>\n",
       "      <td>1979-06-30</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(79)90457-1</td>\n",
       "      <td>55b8337a50d270af02549897d75e22bea89b2083</td>\n",
       "      <td>Epidemic viral gastroenteritis is a significant world wide problem. In developed countries, gast...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21431</th>\n",
       "      <td>i94lyfsh</td>\n",
       "      <td>ddd2ecf42ec86ad66072962081e1ce4594431f9c</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Management of acute and chronic respiratory tract infections</td>\n",
       "      <td>10.1016/0002-9343(88)90456-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3048091</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Pharyngitis, bronchitis, and pneumonia represent the most common respiratory tract infe...</td>\n",
       "      <td>1988-09-16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(88)90456-1</td>\n",
       "      <td>ddd2ecf42ec86ad66072962081e1ce4594431f9c</td>\n",
       "      <td>Respiratory Tract Infections JERROLD J. ELLNER, M.D. Cleveland, CM Pharyngitis, bronchitis, and ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21432</th>\n",
       "      <td>vs5yondw</td>\n",
       "      <td>a55cb4e724091ced46b5e55b982a14525eea1c7e</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Acute bronchitis: Results of U.S. and European trials of antibiotic therapy</td>\n",
       "      <td>10.1016/0002-9343(92)90608-e</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1621745</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract Acute bronchitis, an illness frequently encountered by primary-care physicians, is an i...</td>\n",
       "      <td>1992-06-22</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0002-9343(92)90608-e</td>\n",
       "      <td>a55cb4e724091ced46b5e55b982a14525eea1c7e</td>\n",
       "      <td>A cute bronchitis, an illness frequently encountered by primary-care physicians [1] , is an infl...</td>\n",
       "      <td></td>\n",
       "      <td>Four clinical trials (two in the United States, two in Europe) were initiated to study the relat...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21433</th>\n",
       "      <td>4sbuzcvn</td>\n",
       "      <td>60bf634cf15582a6b7f72eb3811ca5d8b8825c44</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Determination of microsomal lauric acid hydroxylase activity by HPLC with flow-through radiochem...</td>\n",
       "      <td>10.1016/0003-2697(88)90093-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3389520</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract An assay for the microsomal hydroxylation of lauric acid (LA), based on HPLC with flow-...</td>\n",
       "      <td>1988-04-30</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/0003-2697(88)90093-0</td>\n",
       "      <td>60bf634cf15582a6b7f72eb3811ca5d8b8825c44</td>\n",
       "      <td>Laurie acid (LA),' like prostaglandins (PGs) and leukotrienes, is hydroxylated at the w and (w-l...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42967</th>\n",
       "      <td>i2vg8zqc</td>\n",
       "      <td>e9dfbc4084497e74dcaf6bd9c1dbde2b0fb70802</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Clinical Study of Mesenchymal Stem Cell Treatment for Acute Respiratory Distress Syndrome Induce...</td>\n",
       "      <td>10.1016/j.eng.2020.02.006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>Abstract H7N9 viruses quickly spread between mammalian hosts and carry the risk of human-to-huma...</td>\n",
       "      <td>2020-02-28</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.eng.2020.02.006</td>\n",
       "      <td>e9dfbc4084497e74dcaf6bd9c1dbde2b0fb70802</td>\n",
       "      <td>Influenza A viruses (IAVs) are divided into multiple subtypes according to diversified viral sur...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43067</th>\n",
       "      <td>6b7o2cda</td>\n",
       "      <td>f864c42617b8702b776985577444149bbdd7b176</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Clinical features of atypical 2019 novel coronavirus pneumonia with an initially negative RT-PCR...</td>\n",
       "      <td>10.1016/j.jinf.2020.02.008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32092387</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-02-22</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.jinf.2020.02.008</td>\n",
       "      <td>f864c42617b8702b776985577444149bbdd7b176</td>\n",
       "      <td>. Chest CT imaging of the patient (a, c) Chest CT scans obtained at admission show patchy high-d...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43175</th>\n",
       "      <td>87d4r44z</td>\n",
       "      <td>1dda696e7bf65ef55851e1f7778eddcf8a0d8404</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Atypical lung feature on chest CT in a lung adenocarcinoma cancer patient infected with COVID-19</td>\n",
       "      <td>10.1016/j.annonc.2020.03.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-03-09</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.annonc.2020.03.001</td>\n",
       "      <td>1dda696e7bf65ef55851e1f7778eddcf8a0d8404</td>\n",
       "      <td>Atypical lung feature on chest CT in a lung adenocarcinoma cancer patient infected with COVID-19...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43205</th>\n",
       "      <td>0be4wta5</td>\n",
       "      <td>551a07385ce5b146941a9c326abbfa62b56bee5d</td>\n",
       "      <td>Elsevier</td>\n",
       "      <td>Understanding the emerging coronavirus: what it means for health security and infection prevention</td>\n",
       "      <td>10.1016/j.jhin.2020.02.023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32145323</td>\n",
       "      <td>els-covid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-04-30</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://doi.org/10.1016/j.jhin.2020.02.023</td>\n",
       "      <td>551a07385ce5b146941a9c326abbfa62b56bee5d</td>\n",
       "      <td>The current outbreak (COVID-19) of the new coronavirus (SARS-CoV-2) in China and the beginning o...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43346</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Beijing has been one of the epicenters attacked most severely by the SARS-CoV (severe acute resp...</td>\n",
       "      <td>NaT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82e80640af0a98cc07c911c9476f693c5e9a62cf</td>\n",
       "      <td>Accumulated number of probable cases and deaths of SARS in Beijing, China, and the world. Data s...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8267 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cord_uid                                       sha    source  \\\n",
       "21429  zp9k1k3z  aecbc613ebdab36753235197ffb4f35734b5ca63  Elsevier   \n",
       "21430  b10e023s  55b8337a50d270af02549897d75e22bea89b2083  Elsevier   \n",
       "21431  i94lyfsh  ddd2ecf42ec86ad66072962081e1ce4594431f9c  Elsevier   \n",
       "21432  vs5yondw  a55cb4e724091ced46b5e55b982a14525eea1c7e  Elsevier   \n",
       "21433  4sbuzcvn  60bf634cf15582a6b7f72eb3811ca5d8b8825c44  Elsevier   \n",
       "...         ...                                       ...       ...   \n",
       "42967  i2vg8zqc  e9dfbc4084497e74dcaf6bd9c1dbde2b0fb70802  Elsevier   \n",
       "43067  6b7o2cda  f864c42617b8702b776985577444149bbdd7b176  Elsevier   \n",
       "43175  87d4r44z  1dda696e7bf65ef55851e1f7778eddcf8a0d8404  Elsevier   \n",
       "43205  0be4wta5  551a07385ce5b146941a9c326abbfa62b56bee5d  Elsevier   \n",
       "43346       NaN                                       NaN       NaN   \n",
       "\n",
       "                                                                                                     title  \\\n",
       "21429      Clinical and immunologic studies in identical twins discordant for systemic lupus erythematosus   \n",
       "21430                                                                       Epidemic viral gastroenteritis   \n",
       "21431                                         Management of acute and chronic respiratory tract infections   \n",
       "21432                          Acute bronchitis: Results of U.S. and European trials of antibiotic therapy   \n",
       "21433  Determination of microsomal lauric acid hydroxylase activity by HPLC with flow-through radiochem...   \n",
       "...                                                                                                    ...   \n",
       "42967  Clinical Study of Mesenchymal Stem Cell Treatment for Acute Respiratory Distress Syndrome Induce...   \n",
       "43067  Clinical features of atypical 2019 novel coronavirus pneumonia with an initially negative RT-PCR...   \n",
       "43175     Atypical lung feature on chest CT in a lung adenocarcinoma cancer patient infected with COVID-19   \n",
       "43205   Understanding the emerging coronavirus: what it means for health security and infection prevention   \n",
       "43346                                                                                                  NaN   \n",
       "\n",
       "                                doi pmcid pubmed_id    license  \\\n",
       "21429  10.1016/0002-9343(73)90176-9   NaN   4579077  els-covid   \n",
       "21430  10.1016/0002-9343(79)90457-1   NaN    222141  els-covid   \n",
       "21431  10.1016/0002-9343(88)90456-1   NaN   3048091  els-covid   \n",
       "21432  10.1016/0002-9343(92)90608-e   NaN   1621745  els-covid   \n",
       "21433  10.1016/0003-2697(88)90093-0   NaN   3389520  els-covid   \n",
       "...                             ...   ...       ...        ...   \n",
       "42967     10.1016/j.eng.2020.02.006   NaN       NaN  els-covid   \n",
       "43067    10.1016/j.jinf.2020.02.008   NaN  32092387  els-covid   \n",
       "43175  10.1016/j.annonc.2020.03.001   NaN       NaN  els-covid   \n",
       "43205    10.1016/j.jhin.2020.02.023   NaN  32145323  els-covid   \n",
       "43346                           NaN   NaN       NaN        NaN   \n",
       "\n",
       "                                                                                                  abstract  \\\n",
       "21429  Abstract Middle-aged female identical twins, one of whom had systemic lupus erythematosus (SLE),...   \n",
       "21430  Abstract Epidemic viral gastroenteritis is a significant world wide problem. In developed countr...   \n",
       "21431  Abstract Pharyngitis, bronchitis, and pneumonia represent the most common respiratory tract infe...   \n",
       "21432  Abstract Acute bronchitis, an illness frequently encountered by primary-care physicians, is an i...   \n",
       "21433  Abstract An assay for the microsomal hydroxylation of lauric acid (LA), based on HPLC with flow-...   \n",
       "...                                                                                                    ...   \n",
       "42967  Abstract H7N9 viruses quickly spread between mammalian hosts and carry the risk of human-to-huma...   \n",
       "43067                                                                                                  NaN   \n",
       "43175                                                                                                  NaN   \n",
       "43205                                                                                                  NaN   \n",
       "43346  Beijing has been one of the epicenters attacked most severely by the SARS-CoV (severe acute resp...   \n",
       "\n",
       "      publish_time  ... arxiv_id has_pdf_parse has_pmc_xml_parse  \\\n",
       "21429   1973-08-31  ...      NaN          True             False   \n",
       "21430   1979-06-30  ...      NaN          True             False   \n",
       "21431   1988-09-16  ...      NaN          True             False   \n",
       "21432   1992-06-22  ...      NaN          True             False   \n",
       "21433   1988-04-30  ...      NaN          True             False   \n",
       "...            ...  ...      ...           ...               ...   \n",
       "42967   2020-02-28  ...      NaN          True             False   \n",
       "43067   2020-02-22  ...      NaN          True             False   \n",
       "43175   2020-03-09  ...      NaN          True             False   \n",
       "43205   2020-04-30  ...      NaN          True             False   \n",
       "43346          NaT  ...      NaN           NaN               NaN   \n",
       "\n",
       "       full_text_file                                           url  \\\n",
       "21429  custom_license  https://doi.org/10.1016/0002-9343(73)90176-9   \n",
       "21430  custom_license  https://doi.org/10.1016/0002-9343(79)90457-1   \n",
       "21431  custom_license  https://doi.org/10.1016/0002-9343(88)90456-1   \n",
       "21432  custom_license  https://doi.org/10.1016/0002-9343(92)90608-e   \n",
       "21433  custom_license  https://doi.org/10.1016/0003-2697(88)90093-0   \n",
       "...               ...                                           ...   \n",
       "42967  custom_license     https://doi.org/10.1016/j.eng.2020.02.006   \n",
       "43067  custom_license    https://doi.org/10.1016/j.jinf.2020.02.008   \n",
       "43175  custom_license  https://doi.org/10.1016/j.annonc.2020.03.001   \n",
       "43205  custom_license    https://doi.org/10.1016/j.jhin.2020.02.023   \n",
       "43346             NaN                                           NaN   \n",
       "\n",
       "                                   paper_id_pdf  \\\n",
       "21429  aecbc613ebdab36753235197ffb4f35734b5ca63   \n",
       "21430  55b8337a50d270af02549897d75e22bea89b2083   \n",
       "21431  ddd2ecf42ec86ad66072962081e1ce4594431f9c   \n",
       "21432  a55cb4e724091ced46b5e55b982a14525eea1c7e   \n",
       "21433  60bf634cf15582a6b7f72eb3811ca5d8b8825c44   \n",
       "...                                         ...   \n",
       "42967  e9dfbc4084497e74dcaf6bd9c1dbde2b0fb70802   \n",
       "43067  f864c42617b8702b776985577444149bbdd7b176   \n",
       "43175  1dda696e7bf65ef55851e1f7778eddcf8a0d8404   \n",
       "43205  551a07385ce5b146941a9c326abbfa62b56bee5d   \n",
       "43346  82e80640af0a98cc07c911c9476f693c5e9a62cf   \n",
       "\n",
       "                                                                                                 body_text  \\\n",
       "21429  The patient (Fo, ) was a 58 year old mentally retarded white woman, born in a rural area of sout...   \n",
       "21430  Epidemic viral gastroenteritis is a significant world wide problem. In developed countries, gast...   \n",
       "21431  Respiratory Tract Infections JERROLD J. ELLNER, M.D. Cleveland, CM Pharyngitis, bronchitis, and ...   \n",
       "21432  A cute bronchitis, an illness frequently encountered by primary-care physicians [1] , is an infl...   \n",
       "21433  Laurie acid (LA),' like prostaglandins (PGs) and leukotrienes, is hydroxylated at the w and (w-l...   \n",
       "...                                                                                                    ...   \n",
       "42967  Influenza A viruses (IAVs) are divided into multiple subtypes according to diversified viral sur...   \n",
       "43067  . Chest CT imaging of the patient (a, c) Chest CT scans obtained at admission show patchy high-d...   \n",
       "43175  Atypical lung feature on chest CT in a lung adenocarcinoma cancer patient infected with COVID-19...   \n",
       "43205  The current outbreak (COVID-19) of the new coronavirus (SARS-CoV-2) in China and the beginning o...   \n",
       "43346  Accumulated number of probable cases and deaths of SARS in Beijing, China, and the world. Data s...   \n",
       "\n",
       "      methods  \\\n",
       "21429           \n",
       "21430           \n",
       "21431           \n",
       "21432           \n",
       "21433           \n",
       "...       ...   \n",
       "42967           \n",
       "43067           \n",
       "43175           \n",
       "43205           \n",
       "43346           \n",
       "\n",
       "                                                                                                   results  \\\n",
       "21429                                                                                                        \n",
       "21430                                                                                                        \n",
       "21431                                                                                                        \n",
       "21432  Four clinical trials (two in the United States, two in Europe) were initiated to study the relat...   \n",
       "21433                                                                                                        \n",
       "...                                                                                                    ...   \n",
       "42967                                                                                                        \n",
       "43067                                                                                                        \n",
       "43175                                                                                                        \n",
       "43205                                                                                                        \n",
       "43346                                                                                                        \n",
       "\n",
       "      paper_id_pmc  \n",
       "21429          NaN  \n",
       "21430          NaN  \n",
       "21431          NaN  \n",
       "21432          NaN  \n",
       "21433          NaN  \n",
       "...            ...  \n",
       "42967          NaN  \n",
       "43067          NaN  \n",
       "43175          NaN  \n",
       "43205          NaN  \n",
       "43346          NaN  \n",
       "\n",
       "[8267 rows x 24 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: 這個要處理，目前沒有做任何處理。\n",
    "df[df.duplicated('pmcid', keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cord_uid                           1\n",
       "sha                                1\n",
       "source                             1\n",
       "title                             39\n",
       "doi                             1065\n",
       "pmcid                           8267\n",
       "pubmed_id                      13280\n",
       "license                            1\n",
       "abstract                        5051\n",
       "publish_time                       1\n",
       "authors                          697\n",
       "journal                         3720\n",
       "Microsoft Academic Paper ID    42976\n",
       "WHO #Covidence                 42808\n",
       "arxiv_id                       42702\n",
       "has_pdf_parse                      1\n",
       "has_pmc_xml_parse                  1\n",
       "full_text_file                     1\n",
       "url                              121\n",
       "paper_id_pdf                       0\n",
       "body_text                          0\n",
       "methods                            0\n",
       "results                            0\n",
       "paper_id_pmc                   25016\n",
       "dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: 需要再整理前處理過程，目前還沒清乾淨\n",
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some new columns for convenience"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some new columns for convenience\n",
    "df['publish_year'] = df['publish_time'].dt.year#.fillna(-1)\n",
    "# df['link'] = 'http://dx.doi.org/' + df.doi    # dataset now has url column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<regex.Match object; span=(0, 5), match='covid'>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search('COVID', 'covid', re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 43331/43331 [00:15<00:00, 2778.53it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6367"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def contain_covid19_keywords(row):\n",
    "    \n",
    "    covid19_keywords = ['COVID-19', 'covid', 'sar cov 2', 'SARS-CoV-2', '2019-nCov',\n",
    "        '2019 ncov', 'SARS Coronavirus 2', '2019 Novel Coronavirus', 'coronavirus 2019', 'Wuhan coronavirus',\n",
    "        'wuhan pneumonia', 'wuhan virus']\n",
    "    return any(bool(re.search(kw, row['body_text'], re.IGNORECASE)) for kw in covid19_keywords)\n",
    "    \n",
    "df['is_covid19'] = df.progress_apply(contain_covid19_keywords, axis=1)\n",
    "df['is_covid19'].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Language Detection to remove non-english articles and abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.utils import io\n",
    "\n",
    "with io.capture_output() as captured:\n",
    "    !pip install scispacy\n",
    "    !pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_core_sci_lg-0.2.4.tar.gz\n",
    "    !pip install spacy-langdetect\n",
    "    !pip install spac scispacy spacy_langdetect https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.3/en_core_sci_lg-0.2.3.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'scispacy'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-46-fed1b7c47680>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mscispacy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mspacy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0men_core_sci_lg\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mspacy_langdetect\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mLanguageDetector\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'scispacy'"
     ]
    }
   ],
   "source": [
    "import scispacy\n",
    "import spacy\n",
    "import en_core_sci_lg\n",
    "from spacy_langdetect import LanguageDetector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# medium model\n",
    "nlp = en_core_sci_lg.load(disable=[\"tagger\", \"ner\"])\n",
    "nlp.max_length = 2000000\n",
    "nlp.add_pipe(LanguageDetector(), name='language_detector', last=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simple example\n",
    "doc = nlp('This is some English text. Das ist ein Haus. This is a house.')\n",
    "\n",
    "print(doc._.language)\n",
    "print()\n",
    "for s in doc.sents:\n",
    "    print(s._.language)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# example for dataset\n",
    "doc = nlp(df[df['sha'] == '1a8a4dbbaa94ced4ef6af69ec7a09d3fa4c0eece'].body_text.iloc[0])\n",
    "doc[:500]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doc_engl = ''\n",
    "for s in doc.sents:\n",
    "    if (s._.language['language'] == 'en'):\n",
    "        doc_engl += s.text\n",
    "        \n",
    "doc_engl[:2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['body_text'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import langdetect\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "df.loc[21286]\n",
    "結果如下:\n",
    "\n",
    "cord_uid                                                                                           iql8h6td\n",
    "sha                                                                a96599c5c4aeb67f88ec0d9d7f9b5c86b3f37752\n",
    "source                                                                                                  PMC\n",
    "title                                                                    Gammacoronavirus(‡): Coronaviridae\n",
    "doi                                                                            10.1007/978-0-387-95919-1_58\n",
    "pmcid                                                                                            PMC7176155\n",
    "pubmed_id                                                                                               NaN\n",
    "license                                                                                               no-cc\n",
    "abstract                                                                                                NaN\n",
    "publish_time                                                                            2011-01-01 00:00:00\n",
    "authors                                                                                      Decaro, Nicola\n",
    "journal                                                                       The Springer Index of Viruses\n",
    "Microsoft Academic Paper ID                                                                             NaN\n",
    "WHO #Covidence                                                                                          NaN\n",
    "arxiv_id                                                                                                NaN\n",
    "has_pdf_parse                                                                                          True\n",
    "has_pmc_xml_parse                                                                                      True\n",
    "full_text_file                                                                               custom_license\n",
    "url                                                   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7176155/\n",
    "paper_id_pdf                                                       a96599c5c4aeb67f88ec0d9d7f9b5c86b3f37752\n",
    "body_text                      \\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\n",
    "methods_pdf                                                                                                \n",
    "results_pdf                                                                                                \n",
    "paper_id_pmc                                                                                     PMC7176155\n",
    "publish_year                                                                                           2011\n",
    "is_covid19                                                                                            False\n",
    "Name: 21286, dtype: object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "re.search('[A-Za-z]', df.loc[22574, 'body_text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[22560:22570]['body_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def filter_and_detect_language(x):\n",
    "    x = str(x[:2000]).strip()\n",
    "    if re.search('[A-Za-z]', x):\n",
    "        return langdetect.detect(x)\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "# Check language of each text body (only use the first 2000 characters).\n",
    "df['text_language'] = df['body_text'].progress_apply(filter_and_detect_language)\n",
    "# 原本為 df['body_text'].progress_apply(lambda x: nlp(str(x[:2000]))._.language['language']) 慢很多\n",
    "\n",
    "df.text_language.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Number of non-english texts to drop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.loc[df.text_language != 'en'].shape)\n",
    "df = df[df.text_language == 'en']\n",
    "print(df.loc[df.text_language != 'en'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "# Check language of all abstracts\n",
    "\n",
    "# df['abstract_lang'] = df.abstract.apply(lambda x: nlp(str(x))._.language['language'])\n",
    "\n",
    "#  df[df.abstract.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "# Number of non-english abstracts\n",
    "\n",
    "# df[(df.abstract_lang != 'en') & (df.abstract.notnull())].abstract_lang.value_counts()\n",
    "\n",
    "# Keep all english abstracts and those without abstract\n",
    "\n",
    "# df = df[(df.abstract_lang == 'en') | (df.abstract.isnull())]\n",
    "\n",
    "# df.shape\n",
    "\n",
    "# df.paper_id.nunique()\n",
    "\n",
    "# Analyze title/text body of the papers without abstract\n",
    "\n",
    "# temp = df[df.abstract.isnull()].copy()\n",
    "\n",
    "# def remove_non_english_sentences(doc):\n",
    "#     doc = nlp(doc)\n",
    "#     doc_engl = ''\n",
    "#     for s in doc.sents:\n",
    "#         if (s._.language['language'] == 'en'):\n",
    "#             doc_engl += s.text \n",
    "#     return doc_engl\n",
    "\n",
    "# remove_non_english_sentences(df[df.paper_id == '1a8a4dbbaa94ced4ef6af69ec7a09d3fa4c0eece'].body_text.iloc[0])\n",
    "\n",
    "# temp['text_length'] = temp.body_text.apply(lambda x: len(x))\n",
    "\n",
    "# temp['english_text'] = temp.body_text.apply(remove_non_english_sentences)\n",
    "\n",
    "# temp['english_length'] = temp.english_text.apply(lambda x: len(x))\n",
    "\n",
    "# temp.to_csv('df_english.csv', index=False)\n",
    "\n",
    "# (temp.english_length/temp.text_length).hist()\n",
    "\n",
    "# ((temp.english_length/temp.text_length)<0.8).sum()\n",
    "\n",
    "# temp[((temp.english_length/temp.text_length)<0.8)].head()\n",
    "\n",
    "# temp[temp.paper_id == '7925057cfe0cb75ae6079879cb2d22d23e42dfa5'].body_text.values[0][:500]\n",
    "\n",
    "# temp[temp.paper_id == '617197cc751a9208cb0af1b4e31baeddc8d2e985'].body_text.values[0]\n",
    "\n",
    "# temp[temp.paper_id == 'ca51b53fa512085e1aa166d5308602ff1666a90c'].body_text.values[0][:500]\n",
    "\n",
    "# df = df.drop(temp[((temp.english_length/temp.text_length)<0.8)].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "# temp['title_lang'] = df.title.apply(lambda x: nlp(str(x))._.language['language'])\n",
    "\n",
    "# temp.title_lang.value_counts()\n",
    "\n",
    "# Too many false-positves. \n",
    "\n",
    "# temp[temp.paper_id == '6f6b7b1efffae7f3765f29fe801ab63dd35110bb'].body_text.values[0]\n",
    "\n",
    "# temp[temp.title_lang == 'de']\n",
    "\n",
    "# We check the beginning of each text body instead.\n",
    "\n",
    "# temp['text_lang'] = df.body_text.apply(lambda x: nlp(str(x[:2000]))._.language['language'])\n",
    "\n",
    "# temp.text_lang.value_counts()\n",
    "\n",
    "# Number of non-english texts to drop.\n",
    "\n",
    "# df.loc[temp[temp.text_lang != 'en'].index].shape\n",
    "\n",
    "# df = df.drop(temp[temp.text_lang != 'en'].index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract Study Design/ Methodological Keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "# filter_dict = {\n",
    "#     \"discussion\": [\"conclusions\",\"conclusion\",'| discussion', \"discussion\",  'concluding remarks',\n",
    "#                    'discussion and conclusions','conclusion:', 'discussion and conclusion',\n",
    "#                    'conclusions:', 'outcomes', 'conclusions and perspectives', \n",
    "#                    'conclusions and future perspectives', 'conclusions and future directions'],\n",
    "#     \"results\": ['executive summary', 'result', 'summary','results','results and discussion','results:',\n",
    "#                 'comment',\"findings\"],\n",
    "#     \"introduction\": ['introduction', 'background', 'i. introduction','supporting information','| introduction'],\n",
    "#     \"methods\": ['methods','method','statistical methods','materials','materials and methods',\n",
    "#                 'data collection','the study','study design','experimental design','objective',\n",
    "#                 'objectives','procedures','data collection and analysis', 'methodology',\n",
    "#                 'material and methods','the model','experimental procedures','main text',],\n",
    "#     \"statistics\": ['data analysis','statistical analysis', 'analysis','statistical analyses', \n",
    "#                    'statistics','data','measures'],\n",
    "#     \"clinical\": ['diagnosis', 'diagnostic features', \"differential diagnoses\", 'classical signs','prognosis', 'clinical signs', 'pathogenesis',\n",
    "#                  'etiology','differential diagnosis','clinical features', 'case report', 'clinical findings',\n",
    "#                  'clinical presentation'],\n",
    "#     'treatment': ['treatment', 'interventions'],\n",
    "#     \"prevention\": ['epidemiology','risk factors'],\n",
    "#     \"subjects\": ['demographics','samples','subjects', 'study population','control','patients', \n",
    "#                'participants','patient characteristics'],\n",
    "#     \"animals\": ['animals','animal models'],\n",
    "#     \"abstract\": [\"abstract\", 'a b s t r a c t','author summary'], \n",
    "#     \"review\": ['review','literature review','keywords']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "study_designs = {'RCT': ['RCT', 'randomized controlled trial', 'randomised controlled trial', 'randomized control trial', 'randomised control trial',\n",
    "                         'randomized clinical trial','randomised clinical trial'], \n",
    "                'time series analysis': ['time series analysis', 'time series', 'survival analysis'],\n",
    "                'retrospective cohort': ['retrospective cohort'],\n",
    "                'cross-sectional case-control': ['cross-sectional case-control', 'cross sectional case control', 'cross-sectional case control'],\n",
    "                'prospective case-control': ['prospective case-control', 'prospective case control'],\n",
    "                'matched case-control': ['matched case-control', 'matched case control'],\n",
    "                'medical records review': ['medical records review'],\n",
    "                'prevalence survey': ['prevalence survey'],\n",
    "                'syndromic surveillance': ['syndromic surveillance'],\n",
    "                'systematic review': ['systematic review'],\n",
    "                'meta-analysis': ['meta-analysis', 'meta analysis', 'meta-syntheses'],\n",
    "                'interventional study': ['interventional study'],\n",
    "                'association': ['association', 'associated with'],\n",
    "                 'p-value': ['p-value', 'p value'],\n",
    "                 'pseudo-randomized controlled trial': ['pseudo-randomized controlled trial', 'pseudo-randomised controlled trial']\n",
    "                }\n",
    "\n",
    "# Keywords from  [https://docs.google.com/spreadsheets/d/1t2e3CHGxHJBiFgHeW0dfwtvCG4x0CDCzcTFX7yz9Z2E/edit#gid=1217643351](https://docs.google.com/spreadsheets/d/1t2e3CHGxHJBiFgHeW0dfwtvCG4x0CDCzcTFX7yz9Z2E/edit#gid=1217643351)\n",
    "\n",
    "generic_keywords = ['estimation',\n",
    " 'prevalence survey',\n",
    " 'response rate',\n",
    " 'incidence',\n",
    " 'psychometric evaluation of instrument',\n",
    " 'median time to event',\n",
    " 'pooled OR',\n",
    " 'd-pooled',\n",
    " 'randomized controlled trial',\n",
    " 'non-randomized',\n",
    " 'allocation method',\n",
    " 'Cochrane review',\n",
    " 'Cox proportional hazards',\n",
    " 'gamma',\n",
    " 'Weibull',\n",
    " 'pseudo-randomised',\n",
    " 'chart review',\n",
    " 'log odds',\n",
    " 'surveillance',\n",
    " 'time-to-event analysis',\n",
    " 'pooled adjusted odds ratio',\n",
    " 'pooled relative risk',\n",
    " 'data abstraction forms',\n",
    " 'frequency',\n",
    " 'etiology logistic regression',\n",
    " 'exclusion criteria',\n",
    " 'eligibility criteria',\n",
    " 'right-censored',\n",
    " 'pooled odds ratio',\n",
    " 'non-comparative study',\n",
    " 'medical records review',\n",
    " 'CONSORT',\n",
    " 'number of controls per case',\n",
    " 'quasi-randomised',\n",
    " 'risk of bias',\n",
    " 'publication bias',\n",
    " 'syndromic surveillance',\n",
    " 'truncated',\n",
    " 'longitudinal',\n",
    " 'matching criteria',\n",
    " 'double-blind',\n",
    " \"Cohen's d\",\n",
    " 'registry data',\n",
    " 'Adjusted Odds Ratio',\n",
    " 'questionnaire development',\n",
    " 'Kaplan-Meier',\n",
    " 'heterogeneity',\n",
    " 'recruitment',\n",
    " 'randomization method',\n",
    " 'censoring',\n",
    " 'meta-analysis',\n",
    " 'non-randomised',\n",
    " 'β',\n",
    " 'electronic medical records',\n",
    " 'eligibility',\n",
    " 'cross-sectional survey',\n",
    " 'PRISMA',\n",
    " 'prevalence',\n",
    " 'inclusion criteria',\n",
    " 'control arm',\n",
    " 'protocol',\n",
    " 'pooled risk ratio',\n",
    " 'non-response bias',\n",
    " 'baseline',\n",
    " 'retrospective chart review',\n",
    " 'survival analysis',\n",
    " 'logistic regression',\n",
    " 'blind',\n",
    " 'exposure status',\n",
    " 'randomized',\n",
    " 'associated with',\n",
    " 'lognormal',\n",
    " 'systematic review',\n",
    " 'RCT',\n",
    " 'randomised',\n",
    " 'survey instrument',\n",
    " 'interrater reliability',\n",
    " 'randomisation',\n",
    " 'pooled RR',\n",
    " 'hazard ratio',\n",
    " 'AOR',\n",
    " 'potential confounders',\n",
    " 'treatment effect',\n",
    " 'randomized clinical trial',\n",
    " 'data collection instrument',\n",
    " 'pooled AOR',\n",
    " 'association',\n",
    " 'power',\n",
    " \"cohen's kappa\",\n",
    " 'pseudo-randomized',\n",
    " 'treatment arm',\n",
    " 'search string',\n",
    " 'quasi-randomized',\n",
    " 'cohort',\n",
    " 'risk factors',\n",
    " 'difference between means',\n",
    " 'registry',\n",
    " 'inter-rater reliability',\n",
    " 'Odds Ratio',\n",
    " 'placebo',\n",
    " 'databases searched',\n",
    " 'risk factor analysis',\n",
    " 'difference in means',\n",
    " 'random sample',\n",
    " 'etiology',\n",
    " 'i2']\n",
    "\n",
    "# if generic_keyword not in study_designs, then add a design class\n",
    "study_designs_value_list = [value for design in study_designs for value in study_designs[design]]\n",
    "for a in generic_keywords:\n",
    "    if a not in study_designs_value_list:\n",
    "        study_designs[a] = [a]\n",
    "        \n",
    "len([value for design in study_designs for value in study_designs[design]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "_kg_hide-input": true
   },
   "outputs": [],
   "source": [
    "# def tag_study_design(study_designs):\n",
    "#     df['study_design'] = [set() for _ in range(len(df))]\n",
    "#     for tag in study_designs.keys():\n",
    "#         for synonym in study_designs[tag]:\n",
    "#             df[df.abstract.str.contains(synonym, case=False, na=False)].study_design.apply(lambda x: x.add(tag))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tag_study_design(study_designs):\n",
    "    df['study_abstract'] = [set() for _ in range(len(df))]\n",
    "    df['study_methods'] = [set() for _ in range(len(df))]\n",
    "    df['study_results'] = [set() for _ in range(len(df))]\n",
    "\n",
    "    for tag in study_designs.keys():\n",
    "        for synonym in study_designs[tag]:\n",
    "            df[df.abstract.str.contains(synonym, case=False, na=False) | df.title.str.contains(synonym, case=False, na=False)].study_abstract.apply(lambda x: x.add(tag))\n",
    "            df[df.methods.str.contains(synonym, case=False, na=False)].study_methods.apply(lambda x: x.add(tag))\n",
    "            df[df.results.str.contains(synonym, case=False, na=False)].study_results.apply(lambda x: x.add(tag))\n",
    "    \n",
    "    df['study_design'] = df.apply(lambda x: list(x.study_abstract.union(x.study_methods).union(x.study_results)), axis=1)\n",
    "    df.study_abstract = df.study_abstract.apply(lambda x: list(x))\n",
    "    df.study_methods = df.study_abstract.apply(lambda x: list(x))\n",
    "    df.study_results = df.study_results.apply(lambda x: list(x))\n",
    "\n",
    "tag_study_design(study_designs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df.study_design.str.len() != 0].tail(20).study_design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.study_abstract[df.study_abstract.str.len() != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.study_methods[df.study_methods.str.len() != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.study_results[df.study_results.str.len() != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.study_design[df.study_design.str.len() != 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.study_design[(df.study_design.str.len() != 0) & df.is_covid19])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(columns=['cord_uid', 'pmcid', 'pubmed_id', 'full_text_file', 'license', 'text_language',\n",
    "                 'study_abstract', 'study_methods', 'study_results'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export as .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save = 0\n",
    "\n",
    "if save:\n",
    "    df.to_hdf('meta/1-1-create_dataframe.h5', key='data')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
